{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## iGEM labeling functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import nltk\n",
    "import regex as re\n",
    "import ast\n",
    "from collections import defaultdict\n",
    "from snorkel.labeling import labeling_function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test data written by hand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data = [\n",
    "    [[\"carbon\", \"oxygen\"], \"carbon was oxidized by the oxygen\"],\n",
    "    [[\"carbon\", \"oxygen\", \"amino acid\", \"cassie\"], \"the cassie amino acid is connected to the carbon which was oxidixed by the oxygen inside the dna\"],\n",
    "    \n",
    "    [[\"L-GalDH\"], \"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\"],\n",
    "    [[\"L-GalDH\"], \"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\"],\n",
    "    [[\"L-GalL dehydrogenase\"], \"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\"],\n",
    "    [[\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"], \"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### setting up abstain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "ABSTAIN = -1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### labeling functions + small tests written"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_solution_words\n",
    "# If any of the words is in a list of solution terms, we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "solution_terms = [\"buffer\", \"diluent\", \"solute\", \"solvent\", \"saturated\", \"unsaturated\"]\n",
    "@labeling_function()\n",
    "def solution_words(x):\n",
    "    for word in solution_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_mixture_words\n",
    "# If any of the words is in a list of mixture terms, we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "mixture_terms = [\"suspended\", \"mixture\", \"heterogenous\", \"homogeneous\"]\n",
    "@labeling_function()\n",
    "def mixture_words(x):\n",
    "    for word in mixture_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_physical_words\n",
    "# If there are common physical terms, we label FALSE\n",
    "physical_terms = [\"purified\", \"concentration\", \"detection\", \"composed\", \"weight\", \"characteristic\", \"metal\",\n",
    "    \"characterization\", \"color\", \"metalic\", \"characterized\", \"character\", \"consists\", \"pure\", \"compose\", \"assay\"]\n",
    "@labeling_function()\n",
    "def physical_words(x):\n",
    "    for word in physical_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_gene_words\n",
    "# If one of the chemicals is in a list of genetic terms such as reductase, dna, gene, we label FALSE\n",
    "# \n",
    "genetic_terms = [\"reductase\", \"dna\", \"gene\", \"allele\", \"locus\", \"genotype\", \"phenotype\", \"dominant\", \"recessive\", \"additive\", \"phenoset\",\n",
    "    \"diallelic\", \"multiallelic\", \"polyallelic\", \"monomorphic\", \"monoallelic\", \"polymorphism\", \"mutation\", \"complex\", \"trait\", \"multifactorial\",\n",
    "    \"polygenic\", \"monogenic\", \"mixed model\", \"transmission probability\", \"transition probability\", \"epistasis\", \"interaction\", \"pleiotropy\",\n",
    "    \"quantitative trait locus\", \"probit\", \"logit\", \"penetrance\", \"transformation\", \"scale of measurement\", \"identity by descent\", \"identity in state\",\n",
    "    \"haplotype\", \"phase\", \"multilocus\", \"genotype\", \"allelic\", \"association\", \"linkage\", \"disequilibrium\", \"gametic\", \"rna\"]\n",
    "@labeling_function()\n",
    "def gene_words(x):\n",
    "    for word in genetic_terms:\n",
    "        if (word in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_structural_words\n",
    "# If there are common structural terms, we label FALSE\n",
    "structural_terms = [\"loop\", \"sequence\", \"encodes\", \"code\", \"codon\", \"dna\", \"rna\", \"pair\", \"group\", \"active site\",\n",
    "    \"bond\", \"chain\", \"gene\", \"structure\", \"structural\", \"encoding\", \"cdna\"]\n",
    "@labeling_function()\n",
    "def structural_words(x):\n",
    "    for word in structural_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_general_chemical\n",
    "# If any of the words is in a list of common chemical terms such as amino acid, sugar, adenosine, amide, adenine, etc..., we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "chemical_terms = [\"amino acid\", \"sugar\", \"adenosine\", \"amide\", \"adenine\", \"algaecide\", \"amines\", \"base\", \"biocides\",\n",
    "    \"clarifier\", \"hydrocarbon\", \"hydrogenation\", \"molecule\", \"nutrients\", \"polymer\", \"peptide\", \"polypeptide\",\n",
    "    \"tag\", \"functional\", \"residue\", \"activity\", \"enzyme\"]\n",
    "@labeling_function()\n",
    "def general_chemical(x):\n",
    "    for word in chemical_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_group\n",
    "# If there is a close mention of a functional chemical group, we label FALSE (as it is more likely descriptive of a structure than of a reaction)\n",
    "# https://www.masterorganicchemistry.com/2010/10/06/functional-groups-organic-chemistry/\n",
    "common_functional_groups = [\"alkane\", \"alkene\", \"alkyne\", \"benzene ring\", \"phenyl\", \"amine\", \"alcohol\", \"ether\", \"alkyl halide\", \"thiol\",\n",
    "    \"aldehyde\", \"ketone\", \"ester\", \"carboxylic acid\", \"amide\", \"nitrile\", \"epoxide\", \"disulfide\", \"imine\", \"acid chloride\", \"anhydride\", \"nitro\",\n",
    "    \"sulfide\", \"thioether\", \"group\", \"functional\"]\n",
    "@labeling_function()\n",
    "def group(x):\n",
    "    for group in common_functional_groups:\n",
    "        if (group in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_amino_acids\n",
    "# If there is a amino acid mentioned, we label FALSE\n",
    "amino_acids = ['VAL', 'ILE', 'LEU', 'GLU', 'GLN', \\\n",
    "    'ASP', 'ASN' 'HIS' 'TRP', 'PHE', 'TYR',    \\\n",
    "    'ARG', 'LYS', 'SER', 'THR', 'MET', 'ALA',    \\\n",
    "    'GLY', 'PRO', 'CYS', \"amino\"]\n",
    "@labeling_function()\n",
    "def includes_amino_acids(x):\n",
    "    for group in amino_acids:\n",
    "        if (group in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_paper_artifacts\n",
    "# If there are common words from paper headers/footers, we label FALSE\n",
    "common_terms = [\"university\", \"univ\", \"pharma\", \"avenue\", \"street\", \"road\", \"department\", \"usa\", \"reference\", \"ref\",\n",
    "    \"keyword\", \"article\", \"http\", \"png\", \"jpg\"]\n",
    "@labeling_function()\n",
    "def paper_artifacts(x):\n",
    "    for word in common_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_no_terms\n",
    "# If there are words stopping or with no change, we label FALSE\n",
    "common_no_terms = [\"inactivated\", \"unaffected\", \"inactive\", \"inactivates\", \"stops\", \"prevent\", \"inhibit\",\n",
    "    \"denature\"]\n",
    "@labeling_function()\n",
    "def no_terms(x):\n",
    "    for word in common_no_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_adjacent_mentions\n",
    "# If the chemicals are adjacent, we label FALSE\n",
    "@labeling_function()\n",
    "def adjacent_mentions(x):\n",
    "    for index in range(len(x[1])):\n",
    "        for index_2 in range(len(x[1])):\n",
    "            if ((x[1][index] + \" \" + x[1][index_2]) in x[0]):\n",
    "                return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_or\n",
    "# If the chemicals are separated by or, we label FALSE\n",
    "@labeling_function()\n",
    "def sep_or(x):\n",
    "    for index in range(len(x[1])):\n",
    "        for index_2 in range(len(x[1])):\n",
    "            if (x[1][index] + \" or \" + x[1][index_2] in x[0]):\n",
    "                return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_and\n",
    "# If the chemicals are separated by and, we label FALSE\n",
    "@labeling_function()\n",
    "def sep_and(x):\n",
    "    for index in range(len(x[1])):\n",
    "        for index_2 in range(len(x[1])):\n",
    "            if (x[1][index] + \" and \" + x[1][index_2] in x[0]):\n",
    "                return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_comma\n",
    "# If the chemicals are separated by a comma, we label FALSE\n",
    "@labeling_function()\n",
    "def sep_comma(x):\n",
    "    for index in range(len(x[1])):\n",
    "        for index_2 in range(len(x[1])):\n",
    "            if (x[1][index] + \", \" + x[1][index_2] in x[0]):\n",
    "                return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_via\n",
    "# If the chemicals are separated by via, in, etc, we label FALSE\n",
    "common_via_words = [\"via\", \"in\"]\n",
    "@labeling_function()\n",
    "def sep_via(x):\n",
    "    for index in range(len(x[1])):\n",
    "        for index_2 in range(len(x[1])):\n",
    "            for term in common_via_words:\n",
    "                if (x[1][index] + \" \" + term + \" \" + x[1][index_2] in x[0]):\n",
    "                    return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_sym\n",
    "# If the chemicals are separated by a single character, we label FALSE  \n",
    "@labeling_function()\n",
    "def sep_sym(x):\n",
    "    for index in range(len(x[1])):\n",
    "        chem_1_index = x[0].find(x[1][index])\n",
    "        chem_1_len = len(x[1][index])\n",
    "        while (0 <= chem_1_index < len(x[0])):\n",
    "            for index_2 in range(len(x[1])):\n",
    "                chem_2_index = x[0].find(x[1][index_2])\n",
    "                while (0 <= chem_2_index < len(x[0])):\n",
    "                    if (sep_sym_helper(chem_1_index, chem_1_len, chem_2_index)):\n",
    "                        return False\n",
    "                    chem_2_index = x[0].find(x[1][index_2], chem_2_index + 1)\n",
    "            chem_1_index = x[0].find(x[1][index], chem_1_index + 1)\n",
    "    return ABSTAIN\n",
    "\n",
    "def sep_sym_helper(index_1, length, index_2):\n",
    "    if (index_1 + length + 1== index_2):\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# testing separated by a single character (ensuring it works with repeated chemicals)\n",
    "sep_sym([\"cassie is here and her name cassie cas is cas cassie\", [\"cassie\", \"cas\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_followed_ase\n",
    "# If one of the chemicals is followed by a word that ends with -ase, we label FALSE\n",
    "@labeling_function()\n",
    "def followed_ase(x):\n",
    "    sentence = x[0].replace('.', '')\n",
    "    sentence = sentence.replace(',', '')\n",
    "    words = sentence.split(\" \")\n",
    "    for chem in x[1]:\n",
    "        index = -1\n",
    "        while (index < len(words)):\n",
    "            if not(chem in words[index + 1:]):\n",
    "                index = len(words)\n",
    "                break\n",
    "            if (\" \" in chem):\n",
    "                chem_split = chem.split(\" \")\n",
    "                index = words.index(chem_split[-1], index + 1)\n",
    "            else:\n",
    "                index = words.index(chem, index + 1)\n",
    "            if (0 < index + 1 < len(words) and words[index + 1][-3:] == \"ase\"):\n",
    "                return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test followed_ase (making sure it works if the chem is repeated)\n",
    "followed_ase([\"carbon fiber and carbon lactase with oxygen\", [\"carbon\", \"oxygen\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_followed_by_noun\n",
    "# If one of the chemicals is followed by a noun, we label FALSE\n",
    "@labeling_function()\n",
    "def followed_by_noun(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    words = sentence.split(\" \")\n",
    "    tagged = nltk.pos_tag(words)\n",
    "    for chem in x[1]:\n",
    "        index = -1\n",
    "        while (index < len(words)):\n",
    "            if not(chem in words[index + 1:]):\n",
    "                index = len(words)\n",
    "                break\n",
    "            if (\" \" in chem):\n",
    "                chem_split = chem.split(\" \")\n",
    "                index = words.index(chem_split[-1], index + 1)\n",
    "            else:\n",
    "                index = words.index(chem, index + 1)\n",
    "            if (index + 1 < len(words)):\n",
    "                if (tagged[index + 1][1] == \"NN\"):\n",
    "                    return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# testing followed_by_noun (ensuring it works with repeats)\n",
    "followed_by_noun([\"cassie eating apple, eating for cassie table\", [\"cassie\", \"apple\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "False\n",
      "-1\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "# test cases\n",
    "print(followed_by_noun([\"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\", [\"L-GalDH\"]]))\n",
    "print(followed_by_noun([\"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\", [\"L-GalDH\"]]))\n",
    "print(followed_by_noun([\"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\", [\"L-GalL dehydrogenase\"]]))\n",
    "print(followed_by_noun([\"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\", [\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_verb\n",
    "# If the chemicals are separated by a verb, we label TRUE\n",
    "@labeling_function()\n",
    "def sep_verb(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    words = sentence.split(\" \")\n",
    "    indexes = []\n",
    "    for chem in x[1]:\n",
    "        index = -1\n",
    "        while (index < len(words)):\n",
    "            if (\" \" in chem):\n",
    "                chem_split = chem.split(\" \")\n",
    "                if not(chem_split[-1] in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem_split[-1], index + 1)\n",
    "            else:\n",
    "                if not(chem in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem, index + 1)\n",
    "            indexes.append(index)\n",
    "    if not indexes:\n",
    "        return ABSTAIN\n",
    "    min_index = min(indexes)\n",
    "    max_index = max(indexes)\n",
    "    tagged = nltk.pos_tag(words)\n",
    "    for index in range(min_index + 1, max_index):\n",
    "        if (tagged[index][1][:2] == \"VB\"):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# testing sep_verb (works when there are several instances)\n",
    "sep_verb([\"cassie cas cassie hi there oxidized cas\", [\"cassie\", \"cas\"]])\n",
    "sep_verb(['univ sydney, dept pharm, sydney, nsw 2006, australia abstract two distinct mg2+-atpase activities were isolated from triton x-100-solubilized human erythrocyte membranes using a combination of calmodulin-agarose to remove ca2+-atpase and ion exchange chromatography to separate the mg2+-atpase activities',\n",
    " ['ca2+', 'mg2+']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n",
      "-1\n",
      "-1\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# test cases\n",
    "print(sep_verb([\"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\", [\"L-GalDH\"]]))\n",
    "print(sep_verb([\"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\", [\"L-GalDH\"]]))\n",
    "print(sep_verb([\"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\", [\"L-GalL dehydrogenase\"]]))\n",
    "print(sep_verb([\"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\", [\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_adverb\n",
    "# If the chemicals are separated by a adverb, we label FALSE\n",
    "@labeling_function()\n",
    "def sep_adverb(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    words = sentence.split(\" \")\n",
    "    indexes = []\n",
    "    for chem in x[1]:\n",
    "        index = -1\n",
    "        while (index < len(words)):\n",
    "            if (\" \" in chem):\n",
    "                chem_split = chem.split(\" \")\n",
    "                if not(chem_split[-1] in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem_split[-1], index + 1)\n",
    "            else:\n",
    "                if not(chem in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem, index + 1)\n",
    "            indexes.append(index)\n",
    "    if not indexes:\n",
    "        return ABSTAIN\n",
    "    min_index = min(indexes)\n",
    "    max_index = max(indexes)\n",
    "    tagged = nltk.pos_tag(words)\n",
    "    for index in range(min_index + 1, max_index):\n",
    "        if (tagged[index][1][:2] == \"RB\"):\n",
    "            return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# testing sep_adverb (for repeated chemicals)\n",
    "sep_adverb([\"cassie happily danced to the moon with cas\", [\"cassie\", \"cas\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n",
      "-1\n",
      "-1\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "# test cases\n",
    "print(sep_adverb([\"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\", [\"L-GalDH\"]]))\n",
    "print(sep_adverb([\"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\", [\"L-GalDH\"]]))\n",
    "print(sep_adverb([\"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\", [\"L-GalL dehydrogenase\"]]))\n",
    "print(sep_adverb([\"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\", [\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_oxidation_words\n",
    "# If the sentence contains oxidation words, we label True\n",
    "oxidation_words = [\"oxidize\", \"reduce\", \"oxidization\", \"redox\" \"oxidizing\", \"oxidizes\", \"oxidize\", \"reduction\",\n",
    "    \"reduces\", \"oxidise\", \"oxidising\", \"oxide\", \"rust\", \"corrode\"]\n",
    "@labeling_function()\n",
    "def includes_oxidation_words(x):\n",
    "    for word in oxidation_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_combustion_words\n",
    "# If the sentence contains combustion words, we label True\n",
    "combustion_words = [\"combusts\", \"combustion\", \"combust\", \"burn\" \"explode\", \"gas-forming\", \"oxidize\", \"reduction\",\n",
    "    \"reduces\", \"oxidise\", \"oxidising\", \"oxide\"]\n",
    "@labeling_function()\n",
    "def includes_combustion_words(x):\n",
    "    for word in combustion_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_neutralization_words\n",
    "# If the sentence contains neutralization words, we label True\n",
    "neutralization_words = [\"neutralize\", \"neutralization\", \"titrate\", \"titration\", \"buffer\"]\n",
    "@labeling_function()\n",
    "def includes_neutralization_words(x):\n",
    "    for word in neutralization_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_catalyze_words\n",
    "# If the sentence contains catalyze words, we label True\n",
    "catalyze_words = [\"catalyze\", \"catalyst\", \"enzyme\"]\n",
    "@labeling_function()\n",
    "def includes_catalyze_words(x):\n",
    "    for word in catalyze_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_combination_words\n",
    "# If the sentence contains combination words, we label True\n",
    "combination_words = [\"combine\", \"combines\", \"combination\"]\n",
    "@labeling_function()\n",
    "def includes_combination_words(x):\n",
    "    for word in combination_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_decomposition_words\n",
    "# If the sentence contains decomposition words, we label True\n",
    "decomposition_words = [\"decompose\", \"decomposition\"]\n",
    "@labeling_function()\n",
    "def includes_decomposition_words(x):\n",
    "    for word in decomposition_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_replace_words\n",
    "# If the sentence contains decomposition words, we label True\n",
    "replace_words = [\"replaces\", \"replace\", \"replacement\"]\n",
    "@labeling_function()\n",
    "def includes_replace_words(x):\n",
    "    for word in replace_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_reaction_words\n",
    "# If the sentence contains reactions words, we label True\n",
    "reaction_words = [\"convert\", \"yeild\", \"produce\", \"make\", \"react\", \"create\", \"synthesize\", \"conversion\",\n",
    "    \"transformation\", \"transform\", \"synthesise\", \"ferment\", \"agent\",\n",
    "    \"methylaion\", \"displacement\", \"combination\", \"exchange\",\n",
    "    \"precipitate\", \"precipitation\", \"transfer\", \"through\", \"by\", \"produce\", \"activates\"]\n",
    "@labeling_function()\n",
    "def includes_reaction_words(x):\n",
    "    for word in reaction_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_reaction_component_words\n",
    "# If the sentence contains reaction compontents words, we label True\n",
    "reaction_component_words = [\"substrate\", \"product\", \"reactant\"]\n",
    "@labeling_function()\n",
    "def includes_reaction_component_words(x):\n",
    "    for word in reaction_component_words:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_comparison\n",
    "# If the sentence contains reaction compontents words, we label True\n",
    "comparison_words = [\"similar\", \"more\", \"greater\", \"less\", \"increase\", \"decrease\"]\n",
    "@labeling_function()\n",
    "def comparison(x):\n",
    "    for word in comparison_words:\n",
    "        if(word in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_conversion_words\n",
    "# If the sentence contains to, from, into, etc., we label True\n",
    "conversion_words = [\"to\", \"from\", \"into\", \"becomes\", \"became\"]\n",
    "@labeling_function()\n",
    "def sep_converstion_words(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    words = sentence.split(\" \")\n",
    "    indexes = []\n",
    "    for chem in x[1]:\n",
    "        index = -1\n",
    "        while (index < len(words)):\n",
    "            if (\" \" in chem):\n",
    "                chem_split = chem.split(\" \")\n",
    "                if not(chem_split[-1] in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem_split[-1], index + 1)\n",
    "            else:\n",
    "                if not(chem in words[index + 1:]):\n",
    "                    index = len(words)\n",
    "                    break\n",
    "                index = words.index(chem, index + 1)\n",
    "            indexes.append(index)\n",
    "    if not indexes:\n",
    "        return ABSTAIN\n",
    "    min_index = min(indexes)\n",
    "    max_index = max(indexes)\n",
    "    for index in range(min_index + 1, max_index):\n",
    "        if (words[index] in conversion_words):\n",
    "            return True\n",
    "    return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test sep_conversion_words\n",
    "sep_converstion_words([\"cassie truns into cas\", [\"cassie\", \"cas\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### my todo + notes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# things to fix\n",
    "    # create specific reation verb labeling (complete)\n",
    "        # oxidation\n",
    "        # reduction\n",
    "        # combustion\n",
    "        # composition\n",
    "        # decomposition\n",
    "        # etc\n",
    "    # creating more test cases\n",
    "    # current test sentences from\n",
    "        # jacob's csv\n",
    "        # https://academic.oup.com/pcp/article/45/9/1271/1857717\n",
    "    # ways to clean data\n",
    "        # ensure there are at least two chemicals (complete)\n",
    "        # get rid of the greek alphabet (complete)\n",
    "        # get rid of periods and () and ; (complete)\n",
    "        # get rid of repeat chems (complete)\n",
    "    # need to figure out how to deal with things in the sentence twice (complete)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### testing on data written in this file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changed data format so must change written data format before running\n",
    "# for data in test_data:\n",
    "#     print(data)\n",
    "#     print(\"general chemicals\")\n",
    "#     print(general_chemical(data[0]))\n",
    "#     print(\"adjacent mentions\")\n",
    "#     print(adjacent_mentions(data[0], data[1]))\n",
    "#     print(\"sep or\")\n",
    "#     print(sep_or(data[0], data[1]))\n",
    "#     print(\"sep and\")\n",
    "#     print(sep_and(data[0], data[1]))\n",
    "#     print(\"sep comma\")\n",
    "#     print(sep_comma(data[0], data[1]))\n",
    "#     print(\"sep via\")\n",
    "#     print(sep_via(data[0], data[1]))\n",
    "#     print(\"sep sym\")\n",
    "#     print(sep_sym(data[0], data[1]))\n",
    "#     print(\"followed ase\")\n",
    "#     print(followed_ase(data[0], data[1]))\n",
    "#     print(\"group\")\n",
    "#     print(group(data[1]))\n",
    "#     print(\"followed by noun\")\n",
    "#     print(followed_by_noun(data[0], data[1]))\n",
    "#     print(\"sep verb\")\n",
    "#     print(sep_verb(data[0], data[1]))\n",
    "#     print(\"sep adverb\")\n",
    "#     print(sep_adverb(data[0], data[1]))\n",
    "#     print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### importing data in and cleaning it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doc</th>\n",
       "      <th>sentence_index</th>\n",
       "      <th>sentence</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>label</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>genes</th>\n",
       "      <th>substrates</th>\n",
       "      <th>products</th>\n",
       "      <th>enzymes</th>\n",
       "      <th>label index dict</th>\n",
       "      <th>abstract_clean</th>\n",
       "      <th>abstract_expand</th>\n",
       "      <th>abstract_ordered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>12067524</td>\n",
       "      <td>0</td>\n",
       "      <td>BACKGROUND AND AIMS: Glutamic acid decarboxyla...</td>\n",
       "      <td>null, null, null, null, null, null, null, null...</td>\n",
       "      <td>O, O, O, O, B-enzyme, I-enzyme, O, O, B-enzyme...</td>\n",
       "      <td>glutamate, GABA</td>\n",
       "      <td>GAD</td>\n",
       "      <td>glutamate</td>\n",
       "      <td>gamma-aminobutyric, acid, GABA</td>\n",
       "      <td>Glutamic, acid, GAD, EC, 4.1.1.15</td>\n",
       "      <td>{4: 'B-enzyme', 5: 'I-enzyme', 8: 'B-enzyme', ...</td>\n",
       "      <td>BACKGROUND AND AIMS: Glutamic acid decarboxyla...</td>\n",
       "      <td>['These included a specific enzyme activity of...</td>\n",
       "      <td>['BACKGROUND AND AIMS: Glutamic acid decarboxy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>12513997</td>\n",
       "      <td>3</td>\n",
       "      <td>Interestingly, the allele of PRO1 was shown to...</td>\n",
       "      <td>null, null, null, null, null, GENE-Y, null, nu...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, O, O, O, O, B-enzym...</td>\n",
       "      <td>gamma-glutamyl, gamma-glutamyl, L-proline, L-g...</td>\n",
       "      <td>PRO1</td>\n",
       "      <td>L-glutamate</td>\n",
       "      <td>L-proline</td>\n",
       "      <td>gamma-glutamyl, kinase, gamma-glutamyl, phosphate</td>\n",
       "      <td>{13: 'B-enzyme', 14: 'I-enzyme', 16: 'B-enzyme...</td>\n",
       "      <td>We previously isolated a mutant which showed a...</td>\n",
       "      <td>['The approach described in this paper could b...</td>\n",
       "      <td>['We previously isolated a mutant which showed...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>12668769</td>\n",
       "      <td>8</td>\n",
       "      <td>We concluded that FDH has no direct role in th...</td>\n",
       "      <td>null, null, null, GENE-Y, null, null, null, nu...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...</td>\n",
       "      <td>Ser, formate</td>\n",
       "      <td>FDH, FDH</td>\n",
       "      <td>formate</td>\n",
       "      <td>CO</td>\n",
       "      <td>FDH</td>\n",
       "      <td>{23: 'B-SUBSTRATE', 25: 'B-PRODUCT-OF', 31: 'B...</td>\n",
       "      <td>Serine (Ser) biosynthesis in C(3) plants can o...</td>\n",
       "      <td>['In shoots, therefore, the pathway from forma...</td>\n",
       "      <td>['Serine (Ser) biosynthesis in C(3) plants can...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>12742526</td>\n",
       "      <td>6</td>\n",
       "      <td>Even though the activities of MAT and GNMT wer...</td>\n",
       "      <td>null, null, null, null, null, GENE-N, null, GE...</td>\n",
       "      <td>O, O, O, O, O, B-enzyme, O, B-enzyme, O, O, O,...</td>\n",
       "      <td>S-adenosylmethionine, S-adenosylhomocysteine</td>\n",
       "      <td>MAT, GNMT</td>\n",
       "      <td>S-adenosylmethionine</td>\n",
       "      <td>S-adenosylhomocysteine</td>\n",
       "      <td>MAT, GNMT</td>\n",
       "      <td>{5: 'B-enzyme', 7: 'B-enzyme', 15: 'B-SUBSTRAT...</td>\n",
       "      <td>Ames dwarf mice (df/df) are deficient in growt...</td>\n",
       "      <td>['Taken together, the data suggest that methio...</td>\n",
       "      <td>['Ames dwarf mice (df/df) are deficient in gro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>12850267</td>\n",
       "      <td>1</td>\n",
       "      <td>One of the enzymes responsible for the product...</td>\n",
       "      <td>null, null, null, null, null, null, null, null...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, B-PRODUCT-OF, O, B-...</td>\n",
       "      <td>KA, kynurenine, glutamine</td>\n",
       "      <td>KATI, GTK</td>\n",
       "      <td>glutamine</td>\n",
       "      <td>KA, oxoglutaramic, acid</td>\n",
       "      <td>kynurenine, aminotransferase, KATI, GTK, EC, 2...</td>\n",
       "      <td>{9: 'B-PRODUCT-OF', 11: 'B-enzyme', 12: 'I-enz...</td>\n",
       "      <td>Kynurenic acid (KA) is an endogenous glutamate...</td>\n",
       "      <td>['Kynurenic acid (KA) is an endogenous glutama...</td>\n",
       "      <td>['Kynurenic acid (KA) is an endogenous glutama...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>15132128</td>\n",
       "      <td>0</td>\n",
       "      <td>PURPOSE: The fluoropyrimidine carbamate (capec...</td>\n",
       "      <td>null, null, null, null, null, null, CHEMICAL, ...</td>\n",
       "      <td>O, O, O, B-SUBSTRATE, I-SUBSTRATE, O, B-SUBSTR...</td>\n",
       "      <td>capecitabine, 5-fluorouracil, 5-FU, thymidine</td>\n",
       "      <td>TP</td>\n",
       "      <td>fluoropyrimidine, carbamate, capecitabine</td>\n",
       "      <td>5-fluorouracil, 5-FU</td>\n",
       "      <td>thymidine, phosphorylase, TP</td>\n",
       "      <td>{3: 'B-SUBSTRATE', 4: 'I-SUBSTRATE', 6: 'B-SUB...</td>\n",
       "      <td>PURPOSE: The fluoropyrimidine carbamate (capec...</td>\n",
       "      <td>['Favorable enzyme profiles (high TP and low D...</td>\n",
       "      <td>['PURPOSE: The fluoropyrimidine carbamate (cap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>15155769</td>\n",
       "      <td>1</td>\n",
       "      <td>Carnitine acetyltransferases (CrAT) catalyze t...</td>\n",
       "      <td>CHEMICAL, null, null, GENE-Y, null, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...</td>\n",
       "      <td>Carnitine, acetyl-CoA, carnitine, acetylcarnitine</td>\n",
       "      <td>CrAT</td>\n",
       "      <td>acetyl-CoA, carnitine</td>\n",
       "      <td>acetylcarnitine</td>\n",
       "      <td>Carnitine, acetyltransferases, CrAT</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...</td>\n",
       "      <td>In eukaryotes, L-carnitine is involved in ener...</td>\n",
       "      <td>['In eukaryotes, L-carnitine is involved in en...</td>\n",
       "      <td>['In eukaryotes, L-carnitine is involved in en...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>15689518</td>\n",
       "      <td>0</td>\n",
       "      <td>L-serine dehydratase (SDH), a member of the be...</td>\n",
       "      <td>CHEMICAL, null, null, GENE-Y, null, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...</td>\n",
       "      <td>L-serine, L-serine, L-threonine, pyruvate, 2-o...</td>\n",
       "      <td>SDH</td>\n",
       "      <td>L-serine, L-threonine</td>\n",
       "      <td>pyruvate, 2-oxobutyrate</td>\n",
       "      <td>L-serine, dehydratase, SDH</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...</td>\n",
       "      <td>L-serine dehydratase (SDH), a member of the be...</td>\n",
       "      <td>['Furthermore, the activity of hSDH-PLP was as...</td>\n",
       "      <td>['L-serine dehydratase (SDH), a member of the ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>16455797</td>\n",
       "      <td>0</td>\n",
       "      <td>Spermidine/spermine N1-acetyltransferase (SSAT...</td>\n",
       "      <td>null, null, null, GENE-Y, null, null, null, nu...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, O, O, O, O, O, O, O, O,...</td>\n",
       "      <td>polyamine, spermidine, spermine</td>\n",
       "      <td>SSAT</td>\n",
       "      <td>spermidine, spermine</td>\n",
       "      <td>polyamine</td>\n",
       "      <td>Spermidine/spermine, N1-acetyltransferase</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 13: 'B-PRODUCT-...</td>\n",
       "      <td>Spermidine/spermine N1-acetyltransferase (SSAT...</td>\n",
       "      <td>['Spermidine/spermine N1-acetyltransferase (SS...</td>\n",
       "      <td>['Spermidine/spermine N1-acetyltransferase (SS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>16484281</td>\n",
       "      <td>1</td>\n",
       "      <td>Astrocytes may play a role in these manifestat...</td>\n",
       "      <td>null, null, null, null, null, null, null, null...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...</td>\n",
       "      <td>glutamate, glutamine, glutamine</td>\n",
       "      <td>GS</td>\n",
       "      <td>glutamate</td>\n",
       "      <td>glutamine</td>\n",
       "      <td>glutamine, synthetase, GS</td>\n",
       "      <td>{17: 'B-SUBSTRATE', 22: 'B-PRODUCT-OF', 26: 'B...</td>\n",
       "      <td>Excess activation of glutamatergic neurotransm...</td>\n",
       "      <td>['Packing density of GS and GFAP-immunoreactiv...</td>\n",
       "      <td>['Excess activation of glutamatergic neurotran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>16801455</td>\n",
       "      <td>0</td>\n",
       "      <td>Reduced synthesis of nitric oxide (NO) contrib...</td>\n",
       "      <td>null, null, null, null, null, null, CHEMICAL, ...</td>\n",
       "      <td>O, O, O, B-PRODUCT-OF, I-PRODUCT-OF, O, B-PROD...</td>\n",
       "      <td>NO, L-arginine, nitric-oxide</td>\n",
       "      <td>NOS, arginase, arginase</td>\n",
       "      <td>L-arginine</td>\n",
       "      <td>nitric, oxide, NO</td>\n",
       "      <td>constitutive, nitric-oxide, NOS, arginase, I, ...</td>\n",
       "      <td>{3: 'B-PRODUCT-OF', 4: 'I-PRODUCT-OF', 6: 'B-P...</td>\n",
       "      <td>Reduced synthesis of nitric oxide (NO) contrib...</td>\n",
       "      <td>['Reduced synthesis of nitric oxide (NO) contr...</td>\n",
       "      <td>['Reduced synthesis of nitric oxide (NO) contr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>17020418</td>\n",
       "      <td>0</td>\n",
       "      <td>Estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>CHEMICAL, null, null, null, CHEMICAL, null, nu...</td>\n",
       "      <td>B-PRODUCT-OF, O, O, O, B-SUBSTRATE, O, O, B-en...</td>\n",
       "      <td>Estrogens, androgens</td>\n",
       "      <td>CYP450, aromatase</td>\n",
       "      <td>androgens</td>\n",
       "      <td>Estrogens</td>\n",
       "      <td>CYP450, aromatase</td>\n",
       "      <td>{0: 'B-PRODUCT-OF', 4: 'B-SUBSTRATE', 7: 'B-en...</td>\n",
       "      <td>Estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>['These third-generation aromatase inhibitors ...</td>\n",
       "      <td>['Estrogens are biosynthesised from androgens ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>17075268</td>\n",
       "      <td>0</td>\n",
       "      <td>BACKGROUND: Histamine synthesized by histidine...</td>\n",
       "      <td>null, null, CHEMICAL, null, null, CHEMICAL, nu...</td>\n",
       "      <td>O, O, B-PRODUCT-OF, O, O, B-enzyme, I-enzyme, ...</td>\n",
       "      <td>Histamine, histidine, L-histidine</td>\n",
       "      <td>HDC</td>\n",
       "      <td>L-histidine</td>\n",
       "      <td>Histamine</td>\n",
       "      <td>histidine, decarboxylase, HDC</td>\n",
       "      <td>{2: 'B-PRODUCT-OF', 5: 'B-enzyme', 6: 'I-enzym...</td>\n",
       "      <td>BACKGROUND: Histamine synthesized by histidine...</td>\n",
       "      <td>['BACKGROUND: Histamine synthesized by histidi...</td>\n",
       "      <td>['BACKGROUND: Histamine synthesized by histidi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>17079868</td>\n",
       "      <td>1</td>\n",
       "      <td>Methionine synthase reductase (MTRR) is an enz...</td>\n",
       "      <td>CHEMICAL, null, null, null, GENE-Y, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, O, B-enzyme, O, O, O, O...</td>\n",
       "      <td>Methionine, Hcy, methionine</td>\n",
       "      <td>MTRR</td>\n",
       "      <td>Hcy</td>\n",
       "      <td>methionine</td>\n",
       "      <td>Methionine, synthase, MTRR</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 4: 'B-enzyme', ...</td>\n",
       "      <td>Homocysteine (Hcy) is thought to play an impor...</td>\n",
       "      <td>['Homocysteine (Hcy) is thought to play an imp...</td>\n",
       "      <td>['Homocysteine (Hcy) is thought to play an imp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>17210712</td>\n",
       "      <td>6</td>\n",
       "      <td>They all expressed ASS, but not ornithine tran...</td>\n",
       "      <td>null, null, null, GENE-Y, null, null, null, CH...</td>\n",
       "      <td>O, O, O, O, O, O, O, B-enzyme, I-enzyme, O, B-...</td>\n",
       "      <td>ornithine, ornithine, arginine, citrulline, ar...</td>\n",
       "      <td>ASS, OTC, rhArg, ASS</td>\n",
       "      <td>ornithine, arginine, arginine</td>\n",
       "      <td>citrulline</td>\n",
       "      <td>ornithine, transcarbamylase, OTC, rhArg, ASS</td>\n",
       "      <td>{7: 'B-enzyme', 8: 'I-enzyme', 10: 'B-enzyme',...</td>\n",
       "      <td>Hepatocellular carcinoma (HCC) is believed to ...</td>\n",
       "      <td>['We have also shown that the rhArg native enz...</td>\n",
       "      <td>['Hepatocellular carcinoma (HCC) is believed t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>17253779</td>\n",
       "      <td>5</td>\n",
       "      <td>Unlike mouse RetSat (mRetSat), zRetSat A had a...</td>\n",
       "      <td>null, null, GENE-Y, null, GENE-Y, null, null, ...</td>\n",
       "      <td>O, O, O, O, O, O, O, B-enzyme, I-enzyme, O, O,...</td>\n",
       "      <td>all-trans-retinol, all-trans-13,14-dihydroreti...</td>\n",
       "      <td>RetSat, mRetSat</td>\n",
       "      <td>all-trans-retinol</td>\n",
       "      <td>all-trans-13,14-dihydroretinol, all-trans-7,8-...</td>\n",
       "      <td>zRetSat, A</td>\n",
       "      <td>{7: 'B-enzyme', 8: 'I-enzyme', 23: 'B-SUBSTRAT...</td>\n",
       "      <td>Metabolism of vitamin A, all-trans-retinol, le...</td>\n",
       "      <td>['These findings point to a conserved function...</td>\n",
       "      <td>['Metabolism of vitamin A, all-trans-retinol, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>17275903</td>\n",
       "      <td>4</td>\n",
       "      <td>LDH is responsible for pyruvate conversion to ...</td>\n",
       "      <td>GENE-N, null, null, null, CHEMICAL, null, null...</td>\n",
       "      <td>B-enzyme, O, O, O, B-SUBSTRATE, O, O, B-PRODUC...</td>\n",
       "      <td>pyruvate, lactate</td>\n",
       "      <td>LDH</td>\n",
       "      <td>pyruvate</td>\n",
       "      <td>lactate</td>\n",
       "      <td>LDH</td>\n",
       "      <td>{0: 'B-enzyme', 4: 'B-SUBSTRATE', 7: 'B-PRODUC...</td>\n",
       "      <td>The etiology of preeclampsia is unknown but is...</td>\n",
       "      <td>['In this study, we further investigated the r...</td>\n",
       "      <td>['The etiology of preeclampsia is unknown but ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>17284757</td>\n",
       "      <td>1</td>\n",
       "      <td>OBJECTIVE: The aim was to test whether the com...</td>\n",
       "      <td>null, null, null, null, null, null, null, null...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...</td>\n",
       "      <td>PUFA, ALA</td>\n",
       "      <td>FADS2</td>\n",
       "      <td>alpha-linolenic, acid, ALA</td>\n",
       "      <td>PUFA</td>\n",
       "      <td>FADS2</td>\n",
       "      <td>{18: 'B-enzyme', 21: 'B-PRODUCT-OF', 30: 'B-SU...</td>\n",
       "      <td>BACKGROUND: Delta(6)-Desaturase (FADS2) is the...</td>\n",
       "      <td>['Least-squares means from generalized linear ...</td>\n",
       "      <td>['BACKGROUND: Delta(6)-Desaturase (FADS2) is t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>17308037</td>\n",
       "      <td>0</td>\n",
       "      <td>In kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>null, null, null, null, null, null, null, null...</td>\n",
       "      <td>O, O, O, O, O, O, O, O, O, O, O, O, O, O, B-SU...</td>\n",
       "      <td>cAMP, cAMP, AMP, AMP, adenosine</td>\n",
       "      <td>ecto-5'-nucleotidase</td>\n",
       "      <td>cAMP, AMP</td>\n",
       "      <td>AMP, adenosine</td>\n",
       "      <td>ecto-5'-nucleotidase</td>\n",
       "      <td>{14: 'B-SUBSTRATE', 16: 'B-PRODUCT-OF', 23: 'B...</td>\n",
       "      <td>In kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>['Blockade of PDE1 (8-methoxymethyl-3-isobutyl...</td>\n",
       "      <td>[\"In kidneys, stimulation of adenylyl cyclase ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>17929831</td>\n",
       "      <td>0</td>\n",
       "      <td>Adenylosuccinate synthetase (AdSS) catalyzes t...</td>\n",
       "      <td>CHEMICAL, null, null, GENE-Y, null, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, B-enzyme, O, O, O, B-SU...</td>\n",
       "      <td>Adenylosuccinate, Mg2+, aspartate, adenylosucc...</td>\n",
       "      <td>AdSS</td>\n",
       "      <td>Mg2+</td>\n",
       "      <td>adenylosuccinate</td>\n",
       "      <td>Adenylosuccinate, synthetase, AdSS</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...</td>\n",
       "      <td>Adenylosuccinate synthetase (AdSS) catalyzes t...</td>\n",
       "      <td>['In contrast, our studies show that MjAdSS is...</td>\n",
       "      <td>['Adenylosuccinate synthetase (AdSS) catalyzes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>18708991</td>\n",
       "      <td>0</td>\n",
       "      <td>Risperidone is metabolized to its active metab...</td>\n",
       "      <td>CHEMICAL, null, null, null, null, null, null, ...</td>\n",
       "      <td>B-SUBSTRATE, O, O, O, O, O, O, O, B-PRODUCT-OF...</td>\n",
       "      <td>Risperidone, 9-hydroxyrisperidone</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>Risperidone</td>\n",
       "      <td>9-hydroxyrisperidone</td>\n",
       "      <td>cytochrome, P450</td>\n",
       "      <td>{0: 'B-SUBSTRATE', 8: 'B-PRODUCT-OF', 13: 'B-e...</td>\n",
       "      <td>Risperidone is metabolized to its active metab...</td>\n",
       "      <td>['The CYP3A5 genotype did not influence risper...</td>\n",
       "      <td>['Risperidone is metabolized to its active met...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>22872607</td>\n",
       "      <td>0</td>\n",
       "      <td>Aldo-keto reductases (AKRs) metabolize a wide ...</td>\n",
       "      <td>CHEMICAL, null, null, GENE-N, null, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...</td>\n",
       "      <td>Aldo-keto, PAHs, o-quinones, oxygen</td>\n",
       "      <td>AKRs</td>\n",
       "      <td>polycyclic, aromatic, PAHs, o-quinones</td>\n",
       "      <td>oxygen</td>\n",
       "      <td>Aldo-keto, reductases, AKRs</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...</td>\n",
       "      <td>Aldo-keto reductases (AKRs) metabolize a wide ...</td>\n",
       "      <td>['Exposure to PAHs, their metabolites, and ROS...</td>\n",
       "      <td>['Aldo-keto reductases (AKRs) metabolize a wid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>23247009</td>\n",
       "      <td>0</td>\n",
       "      <td>Aldose reductase (AR) catalyzes the reduction ...</td>\n",
       "      <td>CHEMICAL, null, null, GENE-Y, null, null, null...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...</td>\n",
       "      <td>Aldose, aldehydes, alcohol</td>\n",
       "      <td>AR</td>\n",
       "      <td>aldehydes</td>\n",
       "      <td>alcohol</td>\n",
       "      <td>Aldose, reductase, AR</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...</td>\n",
       "      <td>Aldose reductase (AR) catalyzes the reduction ...</td>\n",
       "      <td>['In this study, we found that BGG showed low ...</td>\n",
       "      <td>['Aldose reductase (AR) catalyzes the reductio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>23357976</td>\n",
       "      <td>0</td>\n",
       "      <td>Type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>null, null, CHEMICAL, null, null, null, null, ...</td>\n",
       "      <td>B-enzyme, I-enzyme, O, O, O, O, O, B-enzyme, O...</td>\n",
       "      <td>11β-hydroxysteroid, cortisol, cortisone, glucose</td>\n",
       "      <td>HSD11B2</td>\n",
       "      <td>cortisol</td>\n",
       "      <td>cortisone</td>\n",
       "      <td>Type, 2, HSD11B2</td>\n",
       "      <td>{0: 'B-enzyme', 1: 'I-enzyme', 7: 'B-enzyme', ...</td>\n",
       "      <td>Type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>['Type 2 11beta-hydroxysteroid dehydrogenase e...</td>\n",
       "      <td>['Type 2 11beta-hydroxysteroid dehydrogenase e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>23386702</td>\n",
       "      <td>0</td>\n",
       "      <td>Carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>GENE-N, null, CHEMICAL, null, CHEMICAL, null, ...</td>\n",
       "      <td>B-enzyme, O, B-SUBSTRATE, O, B-SUBSTRATE, O, O...</td>\n",
       "      <td>esters, amides, thioesters, alcohols, amines, ...</td>\n",
       "      <td>Carboxylesterases</td>\n",
       "      <td>esters, amides, thioesters</td>\n",
       "      <td>carboxylic, acids, alcohols, amines, thiols</td>\n",
       "      <td>Carboxylesterases</td>\n",
       "      <td>{0: 'B-enzyme', 2: 'B-SUBSTRATE', 4: 'B-SUBSTR...</td>\n",
       "      <td>Carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>['Objective: This study explores the ability o...</td>\n",
       "      <td>['Carboxylesterases hydrolyze esters, amides, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>23552101</td>\n",
       "      <td>6</td>\n",
       "      <td>PIP5K1B encodes phosphatidylinositol 4-phospha...</td>\n",
       "      <td>GENE-Y, null, null, null, null, null, null, nu...</td>\n",
       "      <td>B-enzyme, O, B-enzyme, I-enzyme, O, O, O, O, O...</td>\n",
       "      <td>phosphatidylinositol-4,5-bisphosphate</td>\n",
       "      <td>PIP5K1B, pip5k1β, actin</td>\n",
       "      <td>phosphatidylinositol, 4-phosphate, PI</td>\n",
       "      <td>phosphatidylinositol-4,5-bisphosphate, PI</td>\n",
       "      <td>PIP5K1B, phosphatidylinositol, 4-phosphate, pi...</td>\n",
       "      <td>{0: 'B-enzyme', 2: 'B-enzyme', 3: 'I-enzyme', ...</td>\n",
       "      <td>Friedreich's ataxia (FRDA) is a progressive ne...</td>\n",
       "      <td>['PIP5K1B encodes phosphatidylinositol 4-phosp...</td>\n",
       "      <td>[\"Friedreich's ataxia (FRDA) is a progressive ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>23578607</td>\n",
       "      <td>4</td>\n",
       "      <td>6-DHSG was metabolised by GSH to form a GSH co...</td>\n",
       "      <td>CHEMICAL, null, null, null, CHEMICAL, null, nu...</td>\n",
       "      <td>B-SUBSTRATE, O, O, O, O, O, O, O, O, O, O, B-P...</td>\n",
       "      <td>6-DHSG, GSH, GSH, GS-6-DHSG</td>\n",
       "      <td>glutathione-S-transferase, GST</td>\n",
       "      <td>6-DHSG</td>\n",
       "      <td>GS-6-DHSG</td>\n",
       "      <td>glutathione-S-transferase, GST</td>\n",
       "      <td>{0: 'B-SUBSTRATE', 11: 'B-PRODUCT-OF', 27: 'B-...</td>\n",
       "      <td>6-Dehydroshogaol (6-DHSG) is a bioactive alpha...</td>\n",
       "      <td>['6-Dehydroshogaol (6-DHSG) is a bioactive alp...</td>\n",
       "      <td>['6-Dehydroshogaol (6-DHSG) is a bioactive alp...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           doc  sentence_index  \\\n",
       "197   12067524               0   \n",
       "252   12513997               3   \n",
       "282   12668769               8   \n",
       "318   12742526               6   \n",
       "323   12850267               1   \n",
       "439   15132128               0   \n",
       "465   15155769               1   \n",
       "484   15689518               0   \n",
       "545   16455797               0   \n",
       "561   16484281               1   \n",
       "607   16801455               0   \n",
       "661   17020418               0   \n",
       "682   17075268               0   \n",
       "694   17079868               1   \n",
       "729   17210712               6   \n",
       "742   17253779               5   \n",
       "759   17275903               4   \n",
       "777   17284757               1   \n",
       "789   17308037               0   \n",
       "881   17929831               0   \n",
       "905   18708991               0   \n",
       "995   22872607               0   \n",
       "1086  23247009               0   \n",
       "1179  23357976               0   \n",
       "1195  23386702               0   \n",
       "1367  23552101               6   \n",
       "1383  23578607               4   \n",
       "\n",
       "                                               sentence  \\\n",
       "197   BACKGROUND AND AIMS: Glutamic acid decarboxyla...   \n",
       "252   Interestingly, the allele of PRO1 was shown to...   \n",
       "282   We concluded that FDH has no direct role in th...   \n",
       "318   Even though the activities of MAT and GNMT wer...   \n",
       "323   One of the enzymes responsible for the product...   \n",
       "439   PURPOSE: The fluoropyrimidine carbamate (capec...   \n",
       "465   Carnitine acetyltransferases (CrAT) catalyze t...   \n",
       "484   L-serine dehydratase (SDH), a member of the be...   \n",
       "545   Spermidine/spermine N1-acetyltransferase (SSAT...   \n",
       "561   Astrocytes may play a role in these manifestat...   \n",
       "607   Reduced synthesis of nitric oxide (NO) contrib...   \n",
       "661   Estrogens are biosynthesised from androgens by...   \n",
       "682   BACKGROUND: Histamine synthesized by histidine...   \n",
       "694   Methionine synthase reductase (MTRR) is an enz...   \n",
       "729   They all expressed ASS, but not ornithine tran...   \n",
       "742   Unlike mouse RetSat (mRetSat), zRetSat A had a...   \n",
       "759   LDH is responsible for pyruvate conversion to ...   \n",
       "777   OBJECTIVE: The aim was to test whether the com...   \n",
       "789   In kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   Adenylosuccinate synthetase (AdSS) catalyzes t...   \n",
       "905   Risperidone is metabolized to its active metab...   \n",
       "995   Aldo-keto reductases (AKRs) metabolize a wide ...   \n",
       "1086  Aldose reductase (AR) catalyzes the reduction ...   \n",
       "1179  Type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  Carboxylesterases hydrolyze esters, amides, an...   \n",
       "1367  PIP5K1B encodes phosphatidylinositol 4-phospha...   \n",
       "1383  6-DHSG was metabolised by GSH to form a GSH co...   \n",
       "\n",
       "                                            entity_type  \\\n",
       "197   null, null, null, null, null, null, null, null...   \n",
       "252   null, null, null, null, null, GENE-Y, null, nu...   \n",
       "282   null, null, null, GENE-Y, null, null, null, nu...   \n",
       "318   null, null, null, null, null, GENE-N, null, GE...   \n",
       "323   null, null, null, null, null, null, null, null...   \n",
       "439   null, null, null, null, null, null, CHEMICAL, ...   \n",
       "465   CHEMICAL, null, null, GENE-Y, null, null, null...   \n",
       "484   CHEMICAL, null, null, GENE-Y, null, null, null...   \n",
       "545   null, null, null, GENE-Y, null, null, null, nu...   \n",
       "561   null, null, null, null, null, null, null, null...   \n",
       "607   null, null, null, null, null, null, CHEMICAL, ...   \n",
       "661   CHEMICAL, null, null, null, CHEMICAL, null, nu...   \n",
       "682   null, null, CHEMICAL, null, null, CHEMICAL, nu...   \n",
       "694   CHEMICAL, null, null, null, GENE-Y, null, null...   \n",
       "729   null, null, null, GENE-Y, null, null, null, CH...   \n",
       "742   null, null, GENE-Y, null, GENE-Y, null, null, ...   \n",
       "759   GENE-N, null, null, null, CHEMICAL, null, null...   \n",
       "777   null, null, null, null, null, null, null, null...   \n",
       "789   null, null, null, null, null, null, null, null...   \n",
       "881   CHEMICAL, null, null, GENE-Y, null, null, null...   \n",
       "905   CHEMICAL, null, null, null, null, null, null, ...   \n",
       "995   CHEMICAL, null, null, GENE-N, null, null, null...   \n",
       "1086  CHEMICAL, null, null, GENE-Y, null, null, null...   \n",
       "1179  null, null, CHEMICAL, null, null, null, null, ...   \n",
       "1195  GENE-N, null, CHEMICAL, null, CHEMICAL, null, ...   \n",
       "1367  GENE-Y, null, null, null, null, null, null, nu...   \n",
       "1383  CHEMICAL, null, null, null, CHEMICAL, null, nu...   \n",
       "\n",
       "                                                  label  \\\n",
       "197   O, O, O, O, B-enzyme, I-enzyme, O, O, B-enzyme...   \n",
       "252   O, O, O, O, O, O, O, O, O, O, O, O, O, B-enzym...   \n",
       "282   O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...   \n",
       "318   O, O, O, O, O, B-enzyme, O, B-enzyme, O, O, O,...   \n",
       "323   O, O, O, O, O, O, O, O, O, B-PRODUCT-OF, O, B-...   \n",
       "439   O, O, O, B-SUBSTRATE, I-SUBSTRATE, O, B-SUBSTR...   \n",
       "465   B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...   \n",
       "484   B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...   \n",
       "545   B-enzyme, I-enzyme, O, O, O, O, O, O, O, O, O,...   \n",
       "561   O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...   \n",
       "607   O, O, O, B-PRODUCT-OF, I-PRODUCT-OF, O, B-PROD...   \n",
       "661   B-PRODUCT-OF, O, O, O, B-SUBSTRATE, O, O, B-en...   \n",
       "682   O, O, B-PRODUCT-OF, O, O, B-enzyme, I-enzyme, ...   \n",
       "694   B-enzyme, I-enzyme, O, O, B-enzyme, O, O, O, O...   \n",
       "729   O, O, O, O, O, O, O, B-enzyme, I-enzyme, O, B-...   \n",
       "742   O, O, O, O, O, O, O, B-enzyme, I-enzyme, O, O,...   \n",
       "759   B-enzyme, O, O, O, B-SUBSTRATE, O, O, B-PRODUC...   \n",
       "777   O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, O...   \n",
       "789   O, O, O, O, O, O, O, O, O, O, O, O, O, O, B-SU...   \n",
       "881   B-enzyme, I-enzyme, O, B-enzyme, O, O, O, B-SU...   \n",
       "905   B-SUBSTRATE, O, O, O, O, O, O, O, B-PRODUCT-OF...   \n",
       "995   B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...   \n",
       "1086  B-enzyme, I-enzyme, O, B-enzyme, O, O, O, O, O...   \n",
       "1179  B-enzyme, I-enzyme, O, O, O, O, O, B-enzyme, O...   \n",
       "1195  B-enzyme, O, B-SUBSTRATE, O, B-SUBSTRATE, O, O...   \n",
       "1367  B-enzyme, O, B-enzyme, I-enzyme, O, O, O, O, O...   \n",
       "1383  B-SUBSTRATE, O, O, O, O, O, O, O, O, O, O, B-P...   \n",
       "\n",
       "                                              chemicals  \\\n",
       "197                                     glutamate, GABA   \n",
       "252   gamma-glutamyl, gamma-glutamyl, L-proline, L-g...   \n",
       "282                                        Ser, formate   \n",
       "318        S-adenosylmethionine, S-adenosylhomocysteine   \n",
       "323                           KA, kynurenine, glutamine   \n",
       "439       capecitabine, 5-fluorouracil, 5-FU, thymidine   \n",
       "465   Carnitine, acetyl-CoA, carnitine, acetylcarnitine   \n",
       "484   L-serine, L-serine, L-threonine, pyruvate, 2-o...   \n",
       "545                     polyamine, spermidine, spermine   \n",
       "561                     glutamate, glutamine, glutamine   \n",
       "607                        NO, L-arginine, nitric-oxide   \n",
       "661                                Estrogens, androgens   \n",
       "682                   Histamine, histidine, L-histidine   \n",
       "694                         Methionine, Hcy, methionine   \n",
       "729   ornithine, ornithine, arginine, citrulline, ar...   \n",
       "742   all-trans-retinol, all-trans-13,14-dihydroreti...   \n",
       "759                                   pyruvate, lactate   \n",
       "777                                           PUFA, ALA   \n",
       "789                     cAMP, cAMP, AMP, AMP, adenosine   \n",
       "881   Adenylosuccinate, Mg2+, aspartate, adenylosucc...   \n",
       "905                   Risperidone, 9-hydroxyrisperidone   \n",
       "995                 Aldo-keto, PAHs, o-quinones, oxygen   \n",
       "1086                         Aldose, aldehydes, alcohol   \n",
       "1179   11β-hydroxysteroid, cortisol, cortisone, glucose   \n",
       "1195  esters, amides, thioesters, alcohols, amines, ...   \n",
       "1367              phosphatidylinositol-4,5-bisphosphate   \n",
       "1383                        6-DHSG, GSH, GSH, GS-6-DHSG   \n",
       "\n",
       "                               genes  \\\n",
       "197                              GAD   \n",
       "252                             PRO1   \n",
       "282                         FDH, FDH   \n",
       "318                        MAT, GNMT   \n",
       "323                        KATI, GTK   \n",
       "439                               TP   \n",
       "465                             CrAT   \n",
       "484                              SDH   \n",
       "545                             SSAT   \n",
       "561                               GS   \n",
       "607          NOS, arginase, arginase   \n",
       "661                CYP450, aromatase   \n",
       "682                              HDC   \n",
       "694                             MTRR   \n",
       "729             ASS, OTC, rhArg, ASS   \n",
       "742                  RetSat, mRetSat   \n",
       "759                              LDH   \n",
       "777                            FADS2   \n",
       "789             ecto-5'-nucleotidase   \n",
       "881                             AdSS   \n",
       "905                           CYP2D6   \n",
       "995                             AKRs   \n",
       "1086                              AR   \n",
       "1179                         HSD11B2   \n",
       "1195               Carboxylesterases   \n",
       "1367         PIP5K1B, pip5k1β, actin   \n",
       "1383  glutathione-S-transferase, GST   \n",
       "\n",
       "                                     substrates  \\\n",
       "197                                   glutamate   \n",
       "252                                 L-glutamate   \n",
       "282                                     formate   \n",
       "318                        S-adenosylmethionine   \n",
       "323                                   glutamine   \n",
       "439   fluoropyrimidine, carbamate, capecitabine   \n",
       "465                       acetyl-CoA, carnitine   \n",
       "484                       L-serine, L-threonine   \n",
       "545                        spermidine, spermine   \n",
       "561                                   glutamate   \n",
       "607                                  L-arginine   \n",
       "661                                   androgens   \n",
       "682                                 L-histidine   \n",
       "694                                         Hcy   \n",
       "729               ornithine, arginine, arginine   \n",
       "742                           all-trans-retinol   \n",
       "759                                    pyruvate   \n",
       "777                  alpha-linolenic, acid, ALA   \n",
       "789                                   cAMP, AMP   \n",
       "881                                        Mg2+   \n",
       "905                                 Risperidone   \n",
       "995      polycyclic, aromatic, PAHs, o-quinones   \n",
       "1086                                  aldehydes   \n",
       "1179                                   cortisol   \n",
       "1195                 esters, amides, thioesters   \n",
       "1367      phosphatidylinositol, 4-phosphate, PI   \n",
       "1383                                     6-DHSG   \n",
       "\n",
       "                                               products  \\\n",
       "197                      gamma-aminobutyric, acid, GABA   \n",
       "252                                           L-proline   \n",
       "282                                                  CO   \n",
       "318                              S-adenosylhomocysteine   \n",
       "323                             KA, oxoglutaramic, acid   \n",
       "439                                5-fluorouracil, 5-FU   \n",
       "465                                     acetylcarnitine   \n",
       "484                             pyruvate, 2-oxobutyrate   \n",
       "545                                           polyamine   \n",
       "561                                           glutamine   \n",
       "607                                   nitric, oxide, NO   \n",
       "661                                           Estrogens   \n",
       "682                                           Histamine   \n",
       "694                                          methionine   \n",
       "729                                          citrulline   \n",
       "742   all-trans-13,14-dihydroretinol, all-trans-7,8-...   \n",
       "759                                             lactate   \n",
       "777                                                PUFA   \n",
       "789                                      AMP, adenosine   \n",
       "881                                    adenylosuccinate   \n",
       "905                                9-hydroxyrisperidone   \n",
       "995                                              oxygen   \n",
       "1086                                            alcohol   \n",
       "1179                                          cortisone   \n",
       "1195        carboxylic, acids, alcohols, amines, thiols   \n",
       "1367          phosphatidylinositol-4,5-bisphosphate, PI   \n",
       "1383                                          GS-6-DHSG   \n",
       "\n",
       "                                                enzymes  \\\n",
       "197                   Glutamic, acid, GAD, EC, 4.1.1.15   \n",
       "252   gamma-glutamyl, kinase, gamma-glutamyl, phosphate   \n",
       "282                                                 FDH   \n",
       "318                                           MAT, GNMT   \n",
       "323   kynurenine, aminotransferase, KATI, GTK, EC, 2...   \n",
       "439                        thymidine, phosphorylase, TP   \n",
       "465                 Carnitine, acetyltransferases, CrAT   \n",
       "484                          L-serine, dehydratase, SDH   \n",
       "545           Spermidine/spermine, N1-acetyltransferase   \n",
       "561                           glutamine, synthetase, GS   \n",
       "607   constitutive, nitric-oxide, NOS, arginase, I, ...   \n",
       "661                                   CYP450, aromatase   \n",
       "682                       histidine, decarboxylase, HDC   \n",
       "694                          Methionine, synthase, MTRR   \n",
       "729        ornithine, transcarbamylase, OTC, rhArg, ASS   \n",
       "742                                          zRetSat, A   \n",
       "759                                                 LDH   \n",
       "777                                               FADS2   \n",
       "789                                ecto-5'-nucleotidase   \n",
       "881                  Adenylosuccinate, synthetase, AdSS   \n",
       "905                                    cytochrome, P450   \n",
       "995                         Aldo-keto, reductases, AKRs   \n",
       "1086                              Aldose, reductase, AR   \n",
       "1179                                   Type, 2, HSD11B2   \n",
       "1195                                  Carboxylesterases   \n",
       "1367  PIP5K1B, phosphatidylinositol, 4-phosphate, pi...   \n",
       "1383                     glutathione-S-transferase, GST   \n",
       "\n",
       "                                       label index dict  \\\n",
       "197   {4: 'B-enzyme', 5: 'I-enzyme', 8: 'B-enzyme', ...   \n",
       "252   {13: 'B-enzyme', 14: 'I-enzyme', 16: 'B-enzyme...   \n",
       "282   {23: 'B-SUBSTRATE', 25: 'B-PRODUCT-OF', 31: 'B...   \n",
       "318   {5: 'B-enzyme', 7: 'B-enzyme', 15: 'B-SUBSTRAT...   \n",
       "323   {9: 'B-PRODUCT-OF', 11: 'B-enzyme', 12: 'I-enz...   \n",
       "439   {3: 'B-SUBSTRATE', 4: 'I-SUBSTRATE', 6: 'B-SUB...   \n",
       "465   {0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...   \n",
       "484   {0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...   \n",
       "545   {0: 'B-enzyme', 1: 'I-enzyme', 13: 'B-PRODUCT-...   \n",
       "561   {17: 'B-SUBSTRATE', 22: 'B-PRODUCT-OF', 26: 'B...   \n",
       "607   {3: 'B-PRODUCT-OF', 4: 'I-PRODUCT-OF', 6: 'B-P...   \n",
       "661   {0: 'B-PRODUCT-OF', 4: 'B-SUBSTRATE', 7: 'B-en...   \n",
       "682   {2: 'B-PRODUCT-OF', 5: 'B-enzyme', 6: 'I-enzym...   \n",
       "694   {0: 'B-enzyme', 1: 'I-enzyme', 4: 'B-enzyme', ...   \n",
       "729   {7: 'B-enzyme', 8: 'I-enzyme', 10: 'B-enzyme',...   \n",
       "742   {7: 'B-enzyme', 8: 'I-enzyme', 23: 'B-SUBSTRAT...   \n",
       "759   {0: 'B-enzyme', 4: 'B-SUBSTRATE', 7: 'B-PRODUC...   \n",
       "777   {18: 'B-enzyme', 21: 'B-PRODUCT-OF', 30: 'B-SU...   \n",
       "789   {14: 'B-SUBSTRATE', 16: 'B-PRODUCT-OF', 23: 'B...   \n",
       "881   {0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...   \n",
       "905   {0: 'B-SUBSTRATE', 8: 'B-PRODUCT-OF', 13: 'B-e...   \n",
       "995   {0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...   \n",
       "1086  {0: 'B-enzyme', 1: 'I-enzyme', 3: 'B-enzyme', ...   \n",
       "1179  {0: 'B-enzyme', 1: 'I-enzyme', 7: 'B-enzyme', ...   \n",
       "1195  {0: 'B-enzyme', 2: 'B-SUBSTRATE', 4: 'B-SUBSTR...   \n",
       "1367  {0: 'B-enzyme', 2: 'B-enzyme', 3: 'I-enzyme', ...   \n",
       "1383  {0: 'B-SUBSTRATE', 11: 'B-PRODUCT-OF', 27: 'B-...   \n",
       "\n",
       "                                         abstract_clean  \\\n",
       "197   BACKGROUND AND AIMS: Glutamic acid decarboxyla...   \n",
       "252   We previously isolated a mutant which showed a...   \n",
       "282   Serine (Ser) biosynthesis in C(3) plants can o...   \n",
       "318   Ames dwarf mice (df/df) are deficient in growt...   \n",
       "323   Kynurenic acid (KA) is an endogenous glutamate...   \n",
       "439   PURPOSE: The fluoropyrimidine carbamate (capec...   \n",
       "465   In eukaryotes, L-carnitine is involved in ener...   \n",
       "484   L-serine dehydratase (SDH), a member of the be...   \n",
       "545   Spermidine/spermine N1-acetyltransferase (SSAT...   \n",
       "561   Excess activation of glutamatergic neurotransm...   \n",
       "607   Reduced synthesis of nitric oxide (NO) contrib...   \n",
       "661   Estrogens are biosynthesised from androgens by...   \n",
       "682   BACKGROUND: Histamine synthesized by histidine...   \n",
       "694   Homocysteine (Hcy) is thought to play an impor...   \n",
       "729   Hepatocellular carcinoma (HCC) is believed to ...   \n",
       "742   Metabolism of vitamin A, all-trans-retinol, le...   \n",
       "759   The etiology of preeclampsia is unknown but is...   \n",
       "777   BACKGROUND: Delta(6)-Desaturase (FADS2) is the...   \n",
       "789   In kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   Adenylosuccinate synthetase (AdSS) catalyzes t...   \n",
       "905   Risperidone is metabolized to its active metab...   \n",
       "995   Aldo-keto reductases (AKRs) metabolize a wide ...   \n",
       "1086  Aldose reductase (AR) catalyzes the reduction ...   \n",
       "1179  Type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  Carboxylesterases hydrolyze esters, amides, an...   \n",
       "1367  Friedreich's ataxia (FRDA) is a progressive ne...   \n",
       "1383  6-Dehydroshogaol (6-DHSG) is a bioactive alpha...   \n",
       "\n",
       "                                        abstract_expand  \\\n",
       "197   ['These included a specific enzyme activity of...   \n",
       "252   ['The approach described in this paper could b...   \n",
       "282   ['In shoots, therefore, the pathway from forma...   \n",
       "318   ['Taken together, the data suggest that methio...   \n",
       "323   ['Kynurenic acid (KA) is an endogenous glutama...   \n",
       "439   ['Favorable enzyme profiles (high TP and low D...   \n",
       "465   ['In eukaryotes, L-carnitine is involved in en...   \n",
       "484   ['Furthermore, the activity of hSDH-PLP was as...   \n",
       "545   ['Spermidine/spermine N1-acetyltransferase (SS...   \n",
       "561   ['Packing density of GS and GFAP-immunoreactiv...   \n",
       "607   ['Reduced synthesis of nitric oxide (NO) contr...   \n",
       "661   ['These third-generation aromatase inhibitors ...   \n",
       "682   ['BACKGROUND: Histamine synthesized by histidi...   \n",
       "694   ['Homocysteine (Hcy) is thought to play an imp...   \n",
       "729   ['We have also shown that the rhArg native enz...   \n",
       "742   ['These findings point to a conserved function...   \n",
       "759   ['In this study, we further investigated the r...   \n",
       "777   ['Least-squares means from generalized linear ...   \n",
       "789   ['Blockade of PDE1 (8-methoxymethyl-3-isobutyl...   \n",
       "881   ['In contrast, our studies show that MjAdSS is...   \n",
       "905   ['The CYP3A5 genotype did not influence risper...   \n",
       "995   ['Exposure to PAHs, their metabolites, and ROS...   \n",
       "1086  ['In this study, we found that BGG showed low ...   \n",
       "1179  ['Type 2 11beta-hydroxysteroid dehydrogenase e...   \n",
       "1195  ['Objective: This study explores the ability o...   \n",
       "1367  ['PIP5K1B encodes phosphatidylinositol 4-phosp...   \n",
       "1383  ['6-Dehydroshogaol (6-DHSG) is a bioactive alp...   \n",
       "\n",
       "                                       abstract_ordered  \n",
       "197   ['BACKGROUND AND AIMS: Glutamic acid decarboxy...  \n",
       "252   ['We previously isolated a mutant which showed...  \n",
       "282   ['Serine (Ser) biosynthesis in C(3) plants can...  \n",
       "318   ['Ames dwarf mice (df/df) are deficient in gro...  \n",
       "323   ['Kynurenic acid (KA) is an endogenous glutama...  \n",
       "439   ['PURPOSE: The fluoropyrimidine carbamate (cap...  \n",
       "465   ['In eukaryotes, L-carnitine is involved in en...  \n",
       "484   ['L-serine dehydratase (SDH), a member of the ...  \n",
       "545   ['Spermidine/spermine N1-acetyltransferase (SS...  \n",
       "561   ['Excess activation of glutamatergic neurotran...  \n",
       "607   ['Reduced synthesis of nitric oxide (NO) contr...  \n",
       "661   ['Estrogens are biosynthesised from androgens ...  \n",
       "682   ['BACKGROUND: Histamine synthesized by histidi...  \n",
       "694   ['Homocysteine (Hcy) is thought to play an imp...  \n",
       "729   ['Hepatocellular carcinoma (HCC) is believed t...  \n",
       "742   ['Metabolism of vitamin A, all-trans-retinol, ...  \n",
       "759   ['The etiology of preeclampsia is unknown but ...  \n",
       "777   ['BACKGROUND: Delta(6)-Desaturase (FADS2) is t...  \n",
       "789   [\"In kidneys, stimulation of adenylyl cyclase ...  \n",
       "881   ['Adenylosuccinate synthetase (AdSS) catalyzes...  \n",
       "905   ['Risperidone is metabolized to its active met...  \n",
       "995   ['Aldo-keto reductases (AKRs) metabolize a wid...  \n",
       "1086  ['Aldose reductase (AR) catalyzes the reductio...  \n",
       "1179  ['Type 2 11beta-hydroxysteroid dehydrogenase e...  \n",
       "1195  ['Carboxylesterases hydrolyze esters, amides, ...  \n",
       "1367  [\"Friedreich's ataxia (FRDA) is a progressive ...  \n",
       "1383  ['6-Dehydroshogaol (6-DHSG) is a bioactive alp...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_df = pd.read_csv(\"sentence_data_cleaned_jacob.csv\")\n",
    "sentence_with_chem_df = sentence_df.dropna()\n",
    "several_chem_df = sentence_with_chem_df.loc[sentence_with_chem_df[\"chemicals\"].str.contains(\",\")]\n",
    "several_chem_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "sentence_chem_df = several_chem_df[[\"sentence\", \"chemicals\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "greek_alphabet = {\n",
    "    u'\\u0393': 'Gamma',\n",
    "    u'\\u0394': 'Delta',\n",
    "    u'\\u0398': 'Theta',\n",
    "    u'\\u039B': 'Lamda',\n",
    "    u'\\u039E': 'Xi',\n",
    "    u'\\u03A0': 'Pi',\n",
    "    u'\\u03A3': 'Sigma',\n",
    "    u'\\u03A6': 'Phi',\n",
    "    u'\\u03A7': 'Chi',\n",
    "    u'\\u03A8': 'Psi',\n",
    "    u'\\u03A9': 'Omega',\n",
    "    u'\\u03B1': 'alpha',\n",
    "    u'\\u03B2': 'beta',\n",
    "    u'\\u03B3': 'gamma',\n",
    "    u'\\u03B4': 'delta',\n",
    "    u'\\u03B5': 'epsilon',\n",
    "    u'\\u03B6': 'zeta',\n",
    "    u'\\u03B7': 'eta',\n",
    "    u'\\u03B8': 'theta',\n",
    "    u'\\u03B9': 'iota',\n",
    "    u'\\u03BA': 'kappa',\n",
    "    u'\\u03BB': 'lamda',\n",
    "    u'\\u03BC': 'mu',\n",
    "    u'\\u03BD': 'nu',\n",
    "    u'\\u03BE': 'xi',\n",
    "    u'\\u03BF': 'omicron',\n",
    "    u'\\u03C0': 'pi',\n",
    "    u'\\u03C1': 'rho',\n",
    "    u'\\u03C3': 'sigma',\n",
    "    u'\\u03C4': 'tau',\n",
    "    u'\\u03C5': 'upsilon',\n",
    "    u'\\u03C6': 'phi',\n",
    "    u'\\u03C7': 'chi',\n",
    "    u'\\u03C8': 'psi',\n",
    "    u'\\u03C9': 'omega'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'β' in greek_alphabet.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_greek(words):\n",
    "    for letter in greek_alphabet.keys():\n",
    "        words = words.replace(letter, greek_alphabet[letter])\n",
    "    return words\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'11beta-hydroxysteroid'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# greek test\n",
    "remove_greek(\"11β-hydroxysteroid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chem_into_array(chemicals):\n",
    "    chemicals = remove_greek(chemicals)\n",
    "    chemicals = chemicals.lower()\n",
    "    sample_chems = chemicals.replace(\"%20\", \" \")\n",
    "    sample_chems_list = sample_chems.split(\", \")\n",
    "    for index in range(len(sample_chems_list)):\n",
    "        while(sample_chems_list[index][-1] == \",\"):\n",
    "            sample_chems_list[index] = sample_chems_list[index][:-1]\n",
    "    sample_chems_list = list(set(sample_chems_list))\n",
    "    return sample_chems_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cassie', 'cas']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# removing excess commas from the end\n",
    "chem_into_array(\"cassie,, cassie, cas\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sentence_cleaned(sentence):\n",
    "    sentence = remove_greek(sentence)\n",
    "    sentence = sentence.lower()\n",
    "    sentence = sentence.replace(\".\", \"\")\n",
    "    sentence = sentence.replace(\")\", \"\")\n",
    "    sentence = sentence.replace(\"(\", \"\")\n",
    "    sentence = sentence.replace(\";\", \"\")\n",
    "    sentence = sentence.replace(\":\", \"\")\n",
    "    return sentence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_6320/3355348865.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>background and aims glutamic acid decarboxylas...</td>\n",
       "      <td>glutamate, GABA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>interestingly, the allele of pro1 was shown to...</td>\n",
       "      <td>gamma-glutamyl, gamma-glutamyl, L-proline, L-g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>we concluded that fdh has no direct role in th...</td>\n",
       "      <td>Ser, formate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>even though the activities of mat and gnmt wer...</td>\n",
       "      <td>S-adenosylmethionine, S-adenosylhomocysteine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>one of the enzymes responsible for the product...</td>\n",
       "      <td>KA, kynurenine, glutamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>purpose the fluoropyrimidine carbamate capecit...</td>\n",
       "      <td>capecitabine, 5-fluorouracil, 5-FU, thymidine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>carnitine acetyltransferases crat catalyze the...</td>\n",
       "      <td>Carnitine, acetyl-CoA, carnitine, acetylcarnitine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>l-serine dehydratase sdh, a member of the beta...</td>\n",
       "      <td>L-serine, L-serine, L-threonine, pyruvate, 2-o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>spermidine/spermine n1-acetyltransferase ssat ...</td>\n",
       "      <td>polyamine, spermidine, spermine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>astrocytes may play a role in these manifestat...</td>\n",
       "      <td>glutamate, glutamine, glutamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>reduced synthesis of nitric oxide no contribut...</td>\n",
       "      <td>NO, L-arginine, nitric-oxide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>Estrogens, androgens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>background histamine synthesized by histidine ...</td>\n",
       "      <td>Histamine, histidine, L-histidine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>methionine synthase reductase mtrr is an enzym...</td>\n",
       "      <td>Methionine, Hcy, methionine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>they all expressed ass, but not ornithine tran...</td>\n",
       "      <td>ornithine, ornithine, arginine, citrulline, ar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>unlike mouse retsat mretsat, zretsat a had an ...</td>\n",
       "      <td>all-trans-retinol, all-trans-13,14-dihydroreti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>ldh is responsible for pyruvate conversion to ...</td>\n",
       "      <td>pyruvate, lactate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>objective the aim was to test whether the comm...</td>\n",
       "      <td>PUFA, ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>in kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>cAMP, cAMP, AMP, AMP, adenosine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>adenylosuccinate synthetase adss catalyzes the...</td>\n",
       "      <td>Adenylosuccinate, Mg2+, aspartate, adenylosucc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>risperidone is metabolized to its active metab...</td>\n",
       "      <td>Risperidone, 9-hydroxyrisperidone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>aldo-keto reductases akrs metabolize a wide ra...</td>\n",
       "      <td>Aldo-keto, PAHs, o-quinones, oxygen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>aldose reductase ar catalyzes the reduction of...</td>\n",
       "      <td>Aldose, aldehydes, alcohol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>11β-hydroxysteroid, cortisol, cortisone, glucose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>esters, amides, thioesters, alcohols, amines, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>pip5k1b encodes phosphatidylinositol 4-phospha...</td>\n",
       "      <td>phosphatidylinositol-4,5-bisphosphate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>6-dhsg was metabolised by gsh to form a gsh co...</td>\n",
       "      <td>6-DHSG, GSH, GSH, GS-6-DHSG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence  \\\n",
       "197   background and aims glutamic acid decarboxylas...   \n",
       "252   interestingly, the allele of pro1 was shown to...   \n",
       "282   we concluded that fdh has no direct role in th...   \n",
       "318   even though the activities of mat and gnmt wer...   \n",
       "323   one of the enzymes responsible for the product...   \n",
       "439   purpose the fluoropyrimidine carbamate capecit...   \n",
       "465   carnitine acetyltransferases crat catalyze the...   \n",
       "484   l-serine dehydratase sdh, a member of the beta...   \n",
       "545   spermidine/spermine n1-acetyltransferase ssat ...   \n",
       "561   astrocytes may play a role in these manifestat...   \n",
       "607   reduced synthesis of nitric oxide no contribut...   \n",
       "661   estrogens are biosynthesised from androgens by...   \n",
       "682   background histamine synthesized by histidine ...   \n",
       "694   methionine synthase reductase mtrr is an enzym...   \n",
       "729   they all expressed ass, but not ornithine tran...   \n",
       "742   unlike mouse retsat mretsat, zretsat a had an ...   \n",
       "759   ldh is responsible for pyruvate conversion to ...   \n",
       "777   objective the aim was to test whether the comm...   \n",
       "789   in kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   adenylosuccinate synthetase adss catalyzes the...   \n",
       "905   risperidone is metabolized to its active metab...   \n",
       "995   aldo-keto reductases akrs metabolize a wide ra...   \n",
       "1086  aldose reductase ar catalyzes the reduction of...   \n",
       "1179  type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  carboxylesterases hydrolyze esters, amides, an...   \n",
       "1367  pip5k1b encodes phosphatidylinositol 4-phospha...   \n",
       "1383  6-dhsg was metabolised by gsh to form a gsh co...   \n",
       "\n",
       "                                              chemicals  \n",
       "197                                     glutamate, GABA  \n",
       "252   gamma-glutamyl, gamma-glutamyl, L-proline, L-g...  \n",
       "282                                        Ser, formate  \n",
       "318        S-adenosylmethionine, S-adenosylhomocysteine  \n",
       "323                           KA, kynurenine, glutamine  \n",
       "439       capecitabine, 5-fluorouracil, 5-FU, thymidine  \n",
       "465   Carnitine, acetyl-CoA, carnitine, acetylcarnitine  \n",
       "484   L-serine, L-serine, L-threonine, pyruvate, 2-o...  \n",
       "545                     polyamine, spermidine, spermine  \n",
       "561                     glutamate, glutamine, glutamine  \n",
       "607                        NO, L-arginine, nitric-oxide  \n",
       "661                                Estrogens, androgens  \n",
       "682                   Histamine, histidine, L-histidine  \n",
       "694                         Methionine, Hcy, methionine  \n",
       "729   ornithine, ornithine, arginine, citrulline, ar...  \n",
       "742   all-trans-retinol, all-trans-13,14-dihydroreti...  \n",
       "759                                   pyruvate, lactate  \n",
       "777                                           PUFA, ALA  \n",
       "789                     cAMP, cAMP, AMP, AMP, adenosine  \n",
       "881   Adenylosuccinate, Mg2+, aspartate, adenylosucc...  \n",
       "905                   Risperidone, 9-hydroxyrisperidone  \n",
       "995                 Aldo-keto, PAHs, o-quinones, oxygen  \n",
       "1086                         Aldose, aldehydes, alcohol  \n",
       "1179   11β-hydroxysteroid, cortisol, cortisone, glucose  \n",
       "1195  esters, amides, thioesters, alcohols, amines, ...  \n",
       "1367              phosphatidylinositol-4,5-bisphosphate  \n",
       "1383                        6-DHSG, GSH, GSH, GS-6-DHSG  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\n",
    "sentence_chem_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_6320/4046753350.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>background and aims glutamic acid decarboxylas...</td>\n",
       "      <td>[glutamate, gaba]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>interestingly, the allele of pro1 was shown to...</td>\n",
       "      <td>[gamma-glutamyl, l-glutamate, l-proline]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>we concluded that fdh has no direct role in th...</td>\n",
       "      <td>[ser, formate]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>even though the activities of mat and gnmt wer...</td>\n",
       "      <td>[s-adenosylhomocysteine, s-adenosylmethionine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>one of the enzymes responsible for the product...</td>\n",
       "      <td>[kynurenine, ka, glutamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>purpose the fluoropyrimidine carbamate capecit...</td>\n",
       "      <td>[thymidine, 5-fu, capecitabine, 5-fluorouracil]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>carnitine acetyltransferases crat catalyze the...</td>\n",
       "      <td>[acetylcarnitine, acetyl-coa, carnitine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>l-serine dehydratase sdh, a member of the beta...</td>\n",
       "      <td>[l-threonine, pyruvate, 2-oxobutyrate, l-serine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>spermidine/spermine n1-acetyltransferase ssat ...</td>\n",
       "      <td>[spermidine, spermine, polyamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>astrocytes may play a role in these manifestat...</td>\n",
       "      <td>[glutamate, glutamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>reduced synthesis of nitric oxide no contribut...</td>\n",
       "      <td>[l-arginine, no, nitric-oxide]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>[androgens, estrogens]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>background histamine synthesized by histidine ...</td>\n",
       "      <td>[histidine, l-histidine, histamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>methionine synthase reductase mtrr is an enzym...</td>\n",
       "      <td>[hcy, methionine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>they all expressed ass, but not ornithine tran...</td>\n",
       "      <td>[citrulline, ornithine, arginine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>unlike mouse retsat mretsat, zretsat a had an ...</td>\n",
       "      <td>[all-trans-7,8-dihydroretinol, all-trans-retin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>ldh is responsible for pyruvate conversion to ...</td>\n",
       "      <td>[pyruvate, lactate]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>objective the aim was to test whether the comm...</td>\n",
       "      <td>[ala, pufa]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>in kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>[adenosine, amp, camp]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>adenylosuccinate synthetase adss catalyzes the...</td>\n",
       "      <td>[adenylosuccinate, mg2+, aspartate, gtp, gdp]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>risperidone is metabolized to its active metab...</td>\n",
       "      <td>[9-hydroxyrisperidone, risperidone]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>aldo-keto reductases akrs metabolize a wide ra...</td>\n",
       "      <td>[o-quinones, oxygen, aldo-keto, pahs]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>aldose reductase ar catalyzes the reduction of...</td>\n",
       "      <td>[aldehydes, alcohol, aldose]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>[cortisone, glucose, cortisol, 11beta-hydroxys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>[thiols, alcohols, amines, thioesters, amides,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>6-dhsg was metabolised by gsh to form a gsh co...</td>\n",
       "      <td>[gsh, 6-dhsg, gs-6-dhsg]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence  \\\n",
       "197   background and aims glutamic acid decarboxylas...   \n",
       "252   interestingly, the allele of pro1 was shown to...   \n",
       "282   we concluded that fdh has no direct role in th...   \n",
       "318   even though the activities of mat and gnmt wer...   \n",
       "323   one of the enzymes responsible for the product...   \n",
       "439   purpose the fluoropyrimidine carbamate capecit...   \n",
       "465   carnitine acetyltransferases crat catalyze the...   \n",
       "484   l-serine dehydratase sdh, a member of the beta...   \n",
       "545   spermidine/spermine n1-acetyltransferase ssat ...   \n",
       "561   astrocytes may play a role in these manifestat...   \n",
       "607   reduced synthesis of nitric oxide no contribut...   \n",
       "661   estrogens are biosynthesised from androgens by...   \n",
       "682   background histamine synthesized by histidine ...   \n",
       "694   methionine synthase reductase mtrr is an enzym...   \n",
       "729   they all expressed ass, but not ornithine tran...   \n",
       "742   unlike mouse retsat mretsat, zretsat a had an ...   \n",
       "759   ldh is responsible for pyruvate conversion to ...   \n",
       "777   objective the aim was to test whether the comm...   \n",
       "789   in kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   adenylosuccinate synthetase adss catalyzes the...   \n",
       "905   risperidone is metabolized to its active metab...   \n",
       "995   aldo-keto reductases akrs metabolize a wide ra...   \n",
       "1086  aldose reductase ar catalyzes the reduction of...   \n",
       "1179  type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  carboxylesterases hydrolyze esters, amides, an...   \n",
       "1383  6-dhsg was metabolised by gsh to form a gsh co...   \n",
       "\n",
       "                                              chemicals  \n",
       "197                                   [glutamate, gaba]  \n",
       "252            [gamma-glutamyl, l-glutamate, l-proline]  \n",
       "282                                      [ser, formate]  \n",
       "318      [s-adenosylhomocysteine, s-adenosylmethionine]  \n",
       "323                         [kynurenine, ka, glutamine]  \n",
       "439     [thymidine, 5-fu, capecitabine, 5-fluorouracil]  \n",
       "465            [acetylcarnitine, acetyl-coa, carnitine]  \n",
       "484    [l-threonine, pyruvate, 2-oxobutyrate, l-serine]  \n",
       "545                   [spermidine, spermine, polyamine]  \n",
       "561                              [glutamate, glutamine]  \n",
       "607                      [l-arginine, no, nitric-oxide]  \n",
       "661                              [androgens, estrogens]  \n",
       "682                 [histidine, l-histidine, histamine]  \n",
       "694                                   [hcy, methionine]  \n",
       "729                   [citrulline, ornithine, arginine]  \n",
       "742   [all-trans-7,8-dihydroretinol, all-trans-retin...  \n",
       "759                                 [pyruvate, lactate]  \n",
       "777                                         [ala, pufa]  \n",
       "789                              [adenosine, amp, camp]  \n",
       "881       [adenylosuccinate, mg2+, aspartate, gtp, gdp]  \n",
       "905                 [9-hydroxyrisperidone, risperidone]  \n",
       "995               [o-quinones, oxygen, aldo-keto, pahs]  \n",
       "1086                       [aldehydes, alcohol, aldose]  \n",
       "1179  [cortisone, glucose, cortisol, 11beta-hydroxys...  \n",
       "1195  [thiols, alcohols, amines, thioesters, amides,...  \n",
       "1383                           [gsh, 6-dhsg, gs-6-dhsg]  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)\n",
    "sentence_chem_df = sentence_chem_df.loc[sentence_chem_df[\"chemicals\"].str.len() > 1]\n",
    "sentence_chem_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_6320/4259470197.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"text\"] = sentence_chem_df[[\"sentence\", \"chemicals\"]].values.tolist()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>background and aims glutamic acid decarboxylas...</td>\n",
       "      <td>[glutamate, gaba]</td>\n",
       "      <td>[background and aims glutamic acid decarboxyla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>interestingly, the allele of pro1 was shown to...</td>\n",
       "      <td>[gamma-glutamyl, l-glutamate, l-proline]</td>\n",
       "      <td>[interestingly, the allele of pro1 was shown t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>we concluded that fdh has no direct role in th...</td>\n",
       "      <td>[ser, formate]</td>\n",
       "      <td>[we concluded that fdh has no direct role in t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>even though the activities of mat and gnmt wer...</td>\n",
       "      <td>[s-adenosylhomocysteine, s-adenosylmethionine]</td>\n",
       "      <td>[even though the activities of mat and gnmt we...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>one of the enzymes responsible for the product...</td>\n",
       "      <td>[kynurenine, ka, glutamine]</td>\n",
       "      <td>[one of the enzymes responsible for the produc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>purpose the fluoropyrimidine carbamate capecit...</td>\n",
       "      <td>[thymidine, 5-fu, capecitabine, 5-fluorouracil]</td>\n",
       "      <td>[purpose the fluoropyrimidine carbamate capeci...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>carnitine acetyltransferases crat catalyze the...</td>\n",
       "      <td>[acetylcarnitine, acetyl-coa, carnitine]</td>\n",
       "      <td>[carnitine acetyltransferases crat catalyze th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>l-serine dehydratase sdh, a member of the beta...</td>\n",
       "      <td>[l-threonine, pyruvate, 2-oxobutyrate, l-serine]</td>\n",
       "      <td>[l-serine dehydratase sdh, a member of the bet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>spermidine/spermine n1-acetyltransferase ssat ...</td>\n",
       "      <td>[spermidine, spermine, polyamine]</td>\n",
       "      <td>[spermidine/spermine n1-acetyltransferase ssat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>astrocytes may play a role in these manifestat...</td>\n",
       "      <td>[glutamate, glutamine]</td>\n",
       "      <td>[astrocytes may play a role in these manifesta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>reduced synthesis of nitric oxide no contribut...</td>\n",
       "      <td>[l-arginine, no, nitric-oxide]</td>\n",
       "      <td>[reduced synthesis of nitric oxide no contribu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>[androgens, estrogens]</td>\n",
       "      <td>[estrogens are biosynthesised from androgens b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>background histamine synthesized by histidine ...</td>\n",
       "      <td>[histidine, l-histidine, histamine]</td>\n",
       "      <td>[background histamine synthesized by histidine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>methionine synthase reductase mtrr is an enzym...</td>\n",
       "      <td>[hcy, methionine]</td>\n",
       "      <td>[methionine synthase reductase mtrr is an enzy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>they all expressed ass, but not ornithine tran...</td>\n",
       "      <td>[citrulline, ornithine, arginine]</td>\n",
       "      <td>[they all expressed ass, but not ornithine tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>unlike mouse retsat mretsat, zretsat a had an ...</td>\n",
       "      <td>[all-trans-7,8-dihydroretinol, all-trans-retin...</td>\n",
       "      <td>[unlike mouse retsat mretsat, zretsat a had an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>ldh is responsible for pyruvate conversion to ...</td>\n",
       "      <td>[pyruvate, lactate]</td>\n",
       "      <td>[ldh is responsible for pyruvate conversion to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>objective the aim was to test whether the comm...</td>\n",
       "      <td>[ala, pufa]</td>\n",
       "      <td>[objective the aim was to test whether the com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>in kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>[adenosine, amp, camp]</td>\n",
       "      <td>[in kidneys, stimulation of adenylyl cyclase c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>adenylosuccinate synthetase adss catalyzes the...</td>\n",
       "      <td>[adenylosuccinate, mg2+, aspartate, gtp, gdp]</td>\n",
       "      <td>[adenylosuccinate synthetase adss catalyzes th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>risperidone is metabolized to its active metab...</td>\n",
       "      <td>[9-hydroxyrisperidone, risperidone]</td>\n",
       "      <td>[risperidone is metabolized to its active meta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>aldo-keto reductases akrs metabolize a wide ra...</td>\n",
       "      <td>[o-quinones, oxygen, aldo-keto, pahs]</td>\n",
       "      <td>[aldo-keto reductases akrs metabolize a wide r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>aldose reductase ar catalyzes the reduction of...</td>\n",
       "      <td>[aldehydes, alcohol, aldose]</td>\n",
       "      <td>[aldose reductase ar catalyzes the reduction o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>[cortisone, glucose, cortisol, 11beta-hydroxys...</td>\n",
       "      <td>[type 2 11beta-hydroxysteroid dehydrogenase en...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>[thiols, alcohols, amines, thioesters, amides,...</td>\n",
       "      <td>[carboxylesterases hydrolyze esters, amides, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>6-dhsg was metabolised by gsh to form a gsh co...</td>\n",
       "      <td>[gsh, 6-dhsg, gs-6-dhsg]</td>\n",
       "      <td>[6-dhsg was metabolised by gsh to form a gsh c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence  \\\n",
       "197   background and aims glutamic acid decarboxylas...   \n",
       "252   interestingly, the allele of pro1 was shown to...   \n",
       "282   we concluded that fdh has no direct role in th...   \n",
       "318   even though the activities of mat and gnmt wer...   \n",
       "323   one of the enzymes responsible for the product...   \n",
       "439   purpose the fluoropyrimidine carbamate capecit...   \n",
       "465   carnitine acetyltransferases crat catalyze the...   \n",
       "484   l-serine dehydratase sdh, a member of the beta...   \n",
       "545   spermidine/spermine n1-acetyltransferase ssat ...   \n",
       "561   astrocytes may play a role in these manifestat...   \n",
       "607   reduced synthesis of nitric oxide no contribut...   \n",
       "661   estrogens are biosynthesised from androgens by...   \n",
       "682   background histamine synthesized by histidine ...   \n",
       "694   methionine synthase reductase mtrr is an enzym...   \n",
       "729   they all expressed ass, but not ornithine tran...   \n",
       "742   unlike mouse retsat mretsat, zretsat a had an ...   \n",
       "759   ldh is responsible for pyruvate conversion to ...   \n",
       "777   objective the aim was to test whether the comm...   \n",
       "789   in kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   adenylosuccinate synthetase adss catalyzes the...   \n",
       "905   risperidone is metabolized to its active metab...   \n",
       "995   aldo-keto reductases akrs metabolize a wide ra...   \n",
       "1086  aldose reductase ar catalyzes the reduction of...   \n",
       "1179  type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  carboxylesterases hydrolyze esters, amides, an...   \n",
       "1383  6-dhsg was metabolised by gsh to form a gsh co...   \n",
       "\n",
       "                                              chemicals  \\\n",
       "197                                   [glutamate, gaba]   \n",
       "252            [gamma-glutamyl, l-glutamate, l-proline]   \n",
       "282                                      [ser, formate]   \n",
       "318      [s-adenosylhomocysteine, s-adenosylmethionine]   \n",
       "323                         [kynurenine, ka, glutamine]   \n",
       "439     [thymidine, 5-fu, capecitabine, 5-fluorouracil]   \n",
       "465            [acetylcarnitine, acetyl-coa, carnitine]   \n",
       "484    [l-threonine, pyruvate, 2-oxobutyrate, l-serine]   \n",
       "545                   [spermidine, spermine, polyamine]   \n",
       "561                              [glutamate, glutamine]   \n",
       "607                      [l-arginine, no, nitric-oxide]   \n",
       "661                              [androgens, estrogens]   \n",
       "682                 [histidine, l-histidine, histamine]   \n",
       "694                                   [hcy, methionine]   \n",
       "729                   [citrulline, ornithine, arginine]   \n",
       "742   [all-trans-7,8-dihydroretinol, all-trans-retin...   \n",
       "759                                 [pyruvate, lactate]   \n",
       "777                                         [ala, pufa]   \n",
       "789                              [adenosine, amp, camp]   \n",
       "881       [adenylosuccinate, mg2+, aspartate, gtp, gdp]   \n",
       "905                 [9-hydroxyrisperidone, risperidone]   \n",
       "995               [o-quinones, oxygen, aldo-keto, pahs]   \n",
       "1086                       [aldehydes, alcohol, aldose]   \n",
       "1179  [cortisone, glucose, cortisol, 11beta-hydroxys...   \n",
       "1195  [thiols, alcohols, amines, thioesters, amides,...   \n",
       "1383                           [gsh, 6-dhsg, gs-6-dhsg]   \n",
       "\n",
       "                                                   text  \n",
       "197   [background and aims glutamic acid decarboxyla...  \n",
       "252   [interestingly, the allele of pro1 was shown t...  \n",
       "282   [we concluded that fdh has no direct role in t...  \n",
       "318   [even though the activities of mat and gnmt we...  \n",
       "323   [one of the enzymes responsible for the produc...  \n",
       "439   [purpose the fluoropyrimidine carbamate capeci...  \n",
       "465   [carnitine acetyltransferases crat catalyze th...  \n",
       "484   [l-serine dehydratase sdh, a member of the bet...  \n",
       "545   [spermidine/spermine n1-acetyltransferase ssat...  \n",
       "561   [astrocytes may play a role in these manifesta...  \n",
       "607   [reduced synthesis of nitric oxide no contribu...  \n",
       "661   [estrogens are biosynthesised from androgens b...  \n",
       "682   [background histamine synthesized by histidine...  \n",
       "694   [methionine synthase reductase mtrr is an enzy...  \n",
       "729   [they all expressed ass, but not ornithine tra...  \n",
       "742   [unlike mouse retsat mretsat, zretsat a had an...  \n",
       "759   [ldh is responsible for pyruvate conversion to...  \n",
       "777   [objective the aim was to test whether the com...  \n",
       "789   [in kidneys, stimulation of adenylyl cyclase c...  \n",
       "881   [adenylosuccinate synthetase adss catalyzes th...  \n",
       "905   [risperidone is metabolized to its active meta...  \n",
       "995   [aldo-keto reductases akrs metabolize a wide r...  \n",
       "1086  [aldose reductase ar catalyzes the reduction o...  \n",
       "1179  [type 2 11beta-hydroxysteroid dehydrogenase en...  \n",
       "1195  [carboxylesterases hydrolyze esters, amides, a...  \n",
       "1383  [6-dhsg was metabolised by gsh to form a gsh c...  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df[\"text\"] = sentence_chem_df[[\"sentence\", \"chemicals\"]].values.tolist()\n",
    "sentence_chem_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### running test on the imported data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['background and aims glutamic acid decarboxylase gad, ec 41115 catalyses the conversion of glutamate to gamma-aminobutyric acid gaba', ['glutamate', 'gaba']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['interestingly, the allele of pro1 was shown to enhance the activities of gamma-glutamyl kinase and gamma-glutamyl phosphate reductase, both of which catalyze the first two steps of l-proline synthesis from l-glutamate and which together may form a complex in vivo', ['gamma-glutamyl', 'l-glutamate', 'l-proline']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['we concluded that fdh has no direct role in the regulation of the above two pathways of ser synthesis the breakdown of formate to co2 by the fdh reaction is the primary and preferred fate of the organic acid in arabidopsis', ['ser', 'formate']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: -1\n",
      "\n",
      "['even though the activities of mat and gnmt were elevated, the concentration of liver s-adenosylmethionine was decreased 24%, p<0001 and s-adenosylhomocysteine increased 113%, p<0001 in the dwarf mice', ['s-adenosylhomocysteine', 's-adenosylmethionine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "sep conversion words: -1\n",
      "\n",
      "['one of the enzymes responsible for the production of ka, kynurenine aminotransferase i kati, also catalyses the reversible transamination of glutamine to oxoglutaramic acid gtk, ec 26115', ['kynurenine', 'ka', 'glutamine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: False\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: True\n",
      "sep conversion words: -1\n",
      "\n",
      "['purpose the fluoropyrimidine carbamate capecitabine is converted to 5-fluorouracil 5-fu by thymidine phosphorylase tp inside target tissues', ['thymidine', '5-fu', 'capecitabine', '5-fluorouracil']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: False\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: False\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['carnitine acetyltransferases crat catalyze the reversible conversion of acetyl-coa and carnitine to acetylcarnitine and free coa', ['acetylcarnitine', 'acetyl-coa', 'carnitine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: False\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['l-serine dehydratase sdh, a member of the beta-family of pyridoxal phosphate-dependent plp enzymes, catalyzes the deamination of l-serine and l-threonine to yield pyruvate or 2-oxobutyrate', ['l-threonine', 'pyruvate', '2-oxobutyrate', 'l-serine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: False\n",
      "sep and: False\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['spermidine/spermine n1-acetyltransferase ssat is a key enzyme in the control of polyamine levels in human cells, as acetylation of spermidine and spermine triggers export or degradation', ['spermidine', 'spermine', 'polyamine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: False\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: False\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: -1\n",
      "\n",
      "['astrocytes may play a role in these manifestations because astrocytes are essential in the regulation of released glutamate and its conversion to glutamine through the enzyme glutamine synthetase gs', ['glutamate', 'glutamine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['reduced synthesis of nitric oxide no contributes to the endothelial dysfunction and may be related to limited availability of l-arginine, the common substrate of constitutive nitric-oxide synthase nos and cytosolic arginase i and mitochondrial arginase ii', ['l-arginine', 'no', 'nitric-oxide']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['estrogens are biosynthesised from androgens by the cyp450 enzyme complex called aromatase', ['androgens', 'estrogens']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['background histamine synthesized by histidine decarboxylase hdc from l-histidine is a major chemical mediator in the development of nasal allergy which is characterized by nasal hypersensitivity', ['histidine', 'l-histidine', 'histamine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['methionine synthase reductase mtrr is an enzyme involved in the conversion of hcy to methionine', ['hcy', 'methionine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['they all expressed ass, but not ornithine transcarbamylase otc, the enzyme that converts ornithine, the product of degradation of arginine with rharg, to citrulline, which is converted back to arginine via ass', ['citrulline', 'ornithine', 'arginine']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['unlike mouse retsat mretsat, zretsat a had an altered bond specificity saturating either the 13-14 or 7-8 double bonds of all-trans-retinol to produce either all-trans-13,14-dihydroretinol or all-trans-7,8-dihydroretinol, respectively', ['all-trans-7,8-dihydroretinol', 'all-trans-retinol', 'all-trans-13,14-dihydroretinol']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: False\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: -1\n",
      "\n",
      "['ldh is responsible for pyruvate conversion to lactate through glycolysis', ['pyruvate', 'lactate']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['objective the aim was to test whether the common deletion [t/-] in the promoter of fads2 affects the pufa biosynthetic pathway and consequently modifies the effect of alpha-linolenic acid ala on myocardial infarction mi', ['ala', 'pufa']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: -1\n",
      "sep conversion words: -1\n",
      "\n",
      "[\"in kidneys, stimulation of adenylyl cyclase causes egress of camp, conversion of camp to amp by ecto-phosphodiesterase, and metabolism of amp to adenosine by ecto-5'-nucleotidase\", ['adenosine', 'amp', 'camp']]\n",
      "general chemicals: False\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['adenylosuccinate synthetase adss catalyzes the mg2+ dependent condensation of a molecule of imp with aspartate to form adenylosuccinate, in a reaction driven by the hydrolysis of gtp to gdp', ['adenylosuccinate', 'mg2+', 'aspartate', 'gtp', 'gdp']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['risperidone is metabolized to its active metabolite, 9-hydroxyrisperidone, mainly by the cytochrome p450 enzymes cyp2d6 and 3a4', ['9-hydroxyrisperidone', 'risperidone']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['aldo-keto reductases akrs metabolize a wide range of substrates, including polycyclic aromatic hydrocarbons pahs, generating metabolites o-quinones and reactive oxygen species ros, which are capable of initiating and promoting carcinogenesis', ['o-quinones', 'oxygen', 'aldo-keto', 'pahs']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: -1\n",
      "\n",
      "['aldose reductase ar catalyzes the reduction of toxic lipid aldehydes to their alcohol products and mediates inflammatory signals triggered by lipopolysaccharide lps', ['aldehydes', 'alcohol', 'aldose']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['type 2 11beta-hydroxysteroid dehydrogenase encoded by the hsd11b2 gene converts cortisol to inactive cortisone, and alteration in this enzymatic activity might affect glucose homeostasis by affecting circulating levels or tissue availability of glucocorticoids', ['cortisone', 'glucose', 'cortisol', '11beta-hydroxysteroid']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['carboxylesterases hydrolyze esters, amides, and thioesters to produce carboxylic acids and resulting alcohols, amines, and thiols, respectively', ['thiols', 'alcohols', 'amines', 'thioesters', 'amides', 'esters']]\n",
      "general chemicals: False\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: False\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "sep conversion words: True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for index in range(len(sentence_chem_df[\"sentence\"])):\n",
    "    arguments = sentence_chem_df[\"text\"].iloc[index]\n",
    "    print(arguments)\n",
    "    print(\"general chemicals: \" + str(general_chemical(arguments)))\n",
    "    print(\"adjacent mentions: \" + str(adjacent_mentions(arguments)))\n",
    "    print(\"sep or: \" + str(sep_or(arguments)))\n",
    "    print(\"sep and: \" + str(sep_and(arguments)))\n",
    "    print(\"sep comma: \" + str(sep_comma(arguments)))\n",
    "    print(\"sep via: \" + str(sep_via(arguments)))\n",
    "    print(\"sep sym: \" + str(sep_sym(arguments)))\n",
    "    print(\"followed ase: \" + str(followed_ase(arguments)))\n",
    "    print(\"group: \" + str(group(arguments)))\n",
    "    print(\"followed by noun: \" + str(followed_by_noun(arguments)))\n",
    "    print(\"sep verb: \" + str(sep_verb(arguments)))\n",
    "    print(\"sep adverb: \" + str(sep_adverb(arguments)))\n",
    "    print(\"includes reaction words: \" + str(includes_reaction_words(arguments)))\n",
    "    print(\"sep conversion words: \" + str(sep_converstion_words(arguments)))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### new csv uncleaned data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lit_id</th>\n",
       "      <th>indices</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sentence</th>\n",
       "      <th>sentence_pos</th>\n",
       "      <th>enzymes</th>\n",
       "      <th>enzyme_locations</th>\n",
       "      <th>chemical_entities_full</th>\n",
       "      <th>chemical_names</th>\n",
       "      <th>chemical_smiles</th>\n",
       "      <th>name_smile_tuples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10.1002/jps.20686</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1297</td>\n",
       "      <td>serial JL 313843 291210 291727 291789 291928 3...</td>\n",
       "      <td>[('serial', 'JJ'), ('JL', 'NN'), ('313843', 'C...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10.1002/jps.20686</td>\n",
       "      <td>1</td>\n",
       "      <td>1298</td>\n",
       "      <td>1324</td>\n",
       "      <td>Published by Elsevier Inc.</td>\n",
       "      <td>[('Published', 'VBN'), ('by', 'IN'), ('Elsevie...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10.1002/jps.20686</td>\n",
       "      <td>2</td>\n",
       "      <td>1325</td>\n",
       "      <td>1345</td>\n",
       "      <td>All rights reserved.</td>\n",
       "      <td>[('All', 'DT'), ('rights', 'NNS'), ('reserved'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10.1002/jps.20686</td>\n",
       "      <td>3</td>\n",
       "      <td>1346</td>\n",
       "      <td>9469</td>\n",
       "      <td>KINETICANALYSESFORSPECIESDIFFERENCESINPGLYCOPR...</td>\n",
       "      <td>[('KINETICANALYSESFORSPECIESDIFFERENCESINPGLYC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Diltiazem', 'start': 1718, 'end': 1...</td>\n",
       "      <td>Diltiazem,, Cyclosporin%20A,, Dexamethasone</td>\n",
       "      <td>COc1ccc(cc1)[C@@H]2Sc3ccccc3N(CCN(C)C)C(=O)[C@...</td>\n",
       "      <td>[('Diltiazem', 'COc1ccc(cc1)[C@@H]2Sc3ccccc3N(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.1002/jps.20686</td>\n",
       "      <td>4</td>\n",
       "      <td>9470</td>\n",
       "      <td>9599</td>\n",
       "      <td>Immunoblot analyses of P-gp expressed in MDR1 ...</td>\n",
       "      <td>[('Immunoblot', 'NN'), ('analyses', 'NNS'), ('...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'H241', 'start': 9594, 'end': 9598, ...</td>\n",
       "      <td>H241</td>\n",
       "      <td>C[C@]12C[C@](O)(C[C@@H]1CC[C@@H]3[C@@H]2CC[C@@...</td>\n",
       "      <td>[('H241', 'C[C@]12C[C@](O)(C[C@@H]1CC[C@@H]3[C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952542</th>\n",
       "      <td>10.1263/jbb.99.623</td>\n",
       "      <td>154</td>\n",
       "      <td>29565</td>\n",
       "      <td>29683</td>\n",
       "      <td>A novel ATP regeneration system using polyphos...</td>\n",
       "      <td>[('A', 'DT'), ('novel', 'JJ'), ('ATP', 'NN'), ...</td>\n",
       "      <td>phosphotransferase, kinase</td>\n",
       "      <td>[('phosphotransferase', 9, 9), ('kinase', 12, ...</td>\n",
       "      <td>[{'text': 'ATP', 'start': 29573, 'end': 29576,...</td>\n",
       "      <td>ATP</td>\n",
       "      <td>Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P](O)(=O)O[...</td>\n",
       "      <td>[('ATP', 'Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952543</th>\n",
       "      <td>10.1263/jbb.99.623</td>\n",
       "      <td>155</td>\n",
       "      <td>29684</td>\n",
       "      <td>29723</td>\n",
       "      <td>91 2001 557 563 20 Fujio T. Maruyama A.</td>\n",
       "      <td>[('91', 'CD'), ('2001', 'CD'), ('557', 'CD'), ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952544</th>\n",
       "      <td>10.1263/jbb.99.623</td>\n",
       "      <td>156</td>\n",
       "      <td>29724</td>\n",
       "      <td>29731</td>\n",
       "      <td>Mori H.</td>\n",
       "      <td>[('Mori', 'NNP'), ('H', 'NNP'), ('.', '.')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'H', 'start': 29729, 'end': 29730, '...</td>\n",
       "      <td>H</td>\n",
       "      <td>[H]</td>\n",
       "      <td>[('H', '[H]')]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952545</th>\n",
       "      <td>10.1263/jbb.99.623</td>\n",
       "      <td>157</td>\n",
       "      <td>29732</td>\n",
       "      <td>29882</td>\n",
       "      <td>Production of useful substances by the couplin...</td>\n",
       "      <td>[('Production', 'NN'), ('of', 'IN'), ('useful'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'ATP', 'start': 29797, 'end': 29800,...</td>\n",
       "      <td>ATP,, ATP</td>\n",
       "      <td>Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P](O)(=O)O[...</td>\n",
       "      <td>[('ATP', 'Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952546</th>\n",
       "      <td>10.1263/jbb.99.623</td>\n",
       "      <td>158</td>\n",
       "      <td>29883</td>\n",
       "      <td>29896</td>\n",
       "      <td>59 2001 83 88</td>\n",
       "      <td>[('59', 'CD'), ('2001', 'CD'), ('83', 'CD'), (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>952547 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    lit_id  indices  start    end  \\\n",
       "0        10.1002/jps.20686        0      0   1297   \n",
       "1        10.1002/jps.20686        1   1298   1324   \n",
       "2        10.1002/jps.20686        2   1325   1345   \n",
       "3        10.1002/jps.20686        3   1346   9469   \n",
       "4        10.1002/jps.20686        4   9470   9599   \n",
       "...                    ...      ...    ...    ...   \n",
       "952542  10.1263/jbb.99.623      154  29565  29683   \n",
       "952543  10.1263/jbb.99.623      155  29684  29723   \n",
       "952544  10.1263/jbb.99.623      156  29724  29731   \n",
       "952545  10.1263/jbb.99.623      157  29732  29882   \n",
       "952546  10.1263/jbb.99.623      158  29883  29896   \n",
       "\n",
       "                                                 sentence  \\\n",
       "0       serial JL 313843 291210 291727 291789 291928 3...   \n",
       "1                              Published by Elsevier Inc.   \n",
       "2                                    All rights reserved.   \n",
       "3       KINETICANALYSESFORSPECIESDIFFERENCESINPGLYCOPR...   \n",
       "4       Immunoblot analyses of P-gp expressed in MDR1 ...   \n",
       "...                                                   ...   \n",
       "952542  A novel ATP regeneration system using polyphos...   \n",
       "952543            91 2001 557 563 20 Fujio T. Maruyama A.   \n",
       "952544                                            Mori H.   \n",
       "952545  Production of useful substances by the couplin...   \n",
       "952546                                      59 2001 83 88   \n",
       "\n",
       "                                             sentence_pos  \\\n",
       "0       [('serial', 'JJ'), ('JL', 'NN'), ('313843', 'C...   \n",
       "1       [('Published', 'VBN'), ('by', 'IN'), ('Elsevie...   \n",
       "2       [('All', 'DT'), ('rights', 'NNS'), ('reserved'...   \n",
       "3       [('KINETICANALYSESFORSPECIESDIFFERENCESINPGLYC...   \n",
       "4       [('Immunoblot', 'NN'), ('analyses', 'NNS'), ('...   \n",
       "...                                                   ...   \n",
       "952542  [('A', 'DT'), ('novel', 'JJ'), ('ATP', 'NN'), ...   \n",
       "952543  [('91', 'CD'), ('2001', 'CD'), ('557', 'CD'), ...   \n",
       "952544        [('Mori', 'NNP'), ('H', 'NNP'), ('.', '.')]   \n",
       "952545  [('Production', 'NN'), ('of', 'IN'), ('useful'...   \n",
       "952546  [('59', 'CD'), ('2001', 'CD'), ('83', 'CD'), (...   \n",
       "\n",
       "                           enzymes  \\\n",
       "0                              NaN   \n",
       "1                              NaN   \n",
       "2                              NaN   \n",
       "3                              NaN   \n",
       "4                              NaN   \n",
       "...                            ...   \n",
       "952542  phosphotransferase, kinase   \n",
       "952543                         NaN   \n",
       "952544                         NaN   \n",
       "952545                         NaN   \n",
       "952546                         NaN   \n",
       "\n",
       "                                         enzyme_locations  \\\n",
       "0                                                      []   \n",
       "1                                                      []   \n",
       "2                                                      []   \n",
       "3                                                      []   \n",
       "4                                                      []   \n",
       "...                                                   ...   \n",
       "952542  [('phosphotransferase', 9, 9), ('kinase', 12, ...   \n",
       "952543                                                 []   \n",
       "952544                                                 []   \n",
       "952545                                                 []   \n",
       "952546                                                 []   \n",
       "\n",
       "                                   chemical_entities_full  \\\n",
       "0                                                      []   \n",
       "1                                                      []   \n",
       "2                                                      []   \n",
       "3       [{'text': 'Diltiazem', 'start': 1718, 'end': 1...   \n",
       "4       [{'text': 'H241', 'start': 9594, 'end': 9598, ...   \n",
       "...                                                   ...   \n",
       "952542  [{'text': 'ATP', 'start': 29573, 'end': 29576,...   \n",
       "952543                                                 []   \n",
       "952544  [{'text': 'H', 'start': 29729, 'end': 29730, '...   \n",
       "952545  [{'text': 'ATP', 'start': 29797, 'end': 29800,...   \n",
       "952546                                                 []   \n",
       "\n",
       "                                     chemical_names  \\\n",
       "0                                               NaN   \n",
       "1                                               NaN   \n",
       "2                                               NaN   \n",
       "3       Diltiazem,, Cyclosporin%20A,, Dexamethasone   \n",
       "4                                              H241   \n",
       "...                                             ...   \n",
       "952542                                          ATP   \n",
       "952543                                          NaN   \n",
       "952544                                            H   \n",
       "952545                                    ATP,, ATP   \n",
       "952546                                          NaN   \n",
       "\n",
       "                                          chemical_smiles  \\\n",
       "0                                                     NaN   \n",
       "1                                                     NaN   \n",
       "2                                                     NaN   \n",
       "3       COc1ccc(cc1)[C@@H]2Sc3ccccc3N(CCN(C)C)C(=O)[C@...   \n",
       "4       C[C@]12C[C@](O)(C[C@@H]1CC[C@@H]3[C@@H]2CC[C@@...   \n",
       "...                                                   ...   \n",
       "952542  Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P](O)(=O)O[...   \n",
       "952543                                                NaN   \n",
       "952544                                                [H]   \n",
       "952545  Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P](O)(=O)O[...   \n",
       "952546                                                NaN   \n",
       "\n",
       "                                        name_smile_tuples  \n",
       "0                                                      []  \n",
       "1                                                      []  \n",
       "2                                                      []  \n",
       "3       [('Diltiazem', 'COc1ccc(cc1)[C@@H]2Sc3ccccc3N(...  \n",
       "4       [('H241', 'C[C@]12C[C@](O)(C[C@@H]1CC[C@@H]3[C...  \n",
       "...                                                   ...  \n",
       "952542  [('ATP', 'Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P...  \n",
       "952543                                                 []  \n",
       "952544                                     [('H', '[H]')]  \n",
       "952545  [('ATP', 'Nc1ncnc2n(cnc12)C3OC(CO[P](O)(=O)O[P...  \n",
       "952546                                                 []  \n",
       "\n",
       "[952547 rows x 12 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uncleaned_df = pd.read_csv(\"../sentence_annotations_elsevier_pmid_split6.csv\")\n",
    "uncleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "uncleaned_no_na_df = uncleaned_df.dropna()\n",
    "uncleaned_several_chem_df = uncleaned_no_na_df.loc[uncleaned_no_na_df[\"chemical_names\"].str.contains(\",\")]\n",
    "uncleaned_several_chem_df = uncleaned_several_chem_df[[\"sentence\", \"chemical_names\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "uncleaned_several_chem_df[\"sentence\"] = uncleaned_several_chem_df[\"sentence\"].apply(sentence_cleaned)\n",
    "uncleaned_several_chem_df[\"chemical_names\"] = uncleaned_several_chem_df[\"chemical_names\"].apply(chem_into_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "150       [calcein-am, cyclosporin a, dexamethasone, dil...\n",
       "271       [sodium dodecyl sulfate, geranyl pyrophosphate...\n",
       "280                                                    [h+]\n",
       "281                                                    [h+]\n",
       "286                                                    [h+]\n",
       "                                ...                        \n",
       "952414                    [vancomycin, d-alanine-d-alanine]\n",
       "952450                         [tetrahydrofurane, methanol]\n",
       "952517                                      [sucrose, kato]\n",
       "952526                     [d-alanyl-d-alanine, vancomycin]\n",
       "952541                                [adenylate, amp, atp]\n",
       "Name: chemical_names, Length: 50212, dtype: object"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uncleaned_several_chem_df[\"chemical_names\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "re_check_several_chem_uncleaned_df = uncleaned_several_chem_df.loc[uncleaned_several_chem_df[\"chemical_names\"].str.len() > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemical_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>we selected diltiazem, cyclosporin a, and dexa...</td>\n",
       "      <td>[calcein-am, cyclosporin a, dexamethasone, dil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>croteau r washington state univ, inst biol che...</td>\n",
       "      <td>[sodium dodecyl sulfate, geranyl pyrophosphate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>chemicalmodificationschickenliverpyruvatecarbo...</td>\n",
       "      <td>[lysine, pyruvate, cysteine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>ash de temple univ, hlth sci ctr, sch med, dep...</td>\n",
       "      <td>[pyruvate, cysteine, lysine, o-phthalaldehyde,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>at a one- to two-fold molar excess over active...</td>\n",
       "      <td>[pyruvate, oxaloacetate, adp]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952414</th>\n",
       "      <td>key words d-alanine-d-alanine ligase d-amino a...</td>\n",
       "      <td>[vancomycin, d-alanine-d-alanine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952450</th>\n",
       "      <td>the mobile phase consisted of a linear gradien...</td>\n",
       "      <td>[tetrahydrofurane, methanol]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952517</th>\n",
       "      <td>weber v falkenhagen d subpol a novel sucrose-b...</td>\n",
       "      <td>[sucrose, kato]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952526</th>\n",
       "      <td>active-site mutants of the vanc2 d-alanyl-d-se...</td>\n",
       "      <td>[d-alanyl-d-alanine, vancomycin]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952541</th>\n",
       "      <td>in vitro atp regeneration from polyphosphate a...</td>\n",
       "      <td>[adenylate, amp, atp]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44724 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 sentence  \\\n",
       "150     we selected diltiazem, cyclosporin a, and dexa...   \n",
       "271     croteau r washington state univ, inst biol che...   \n",
       "295     chemicalmodificationschickenliverpyruvatecarbo...   \n",
       "296     ash de temple univ, hlth sci ctr, sch med, dep...   \n",
       "298     at a one- to two-fold molar excess over active...   \n",
       "...                                                   ...   \n",
       "952414  key words d-alanine-d-alanine ligase d-amino a...   \n",
       "952450  the mobile phase consisted of a linear gradien...   \n",
       "952517  weber v falkenhagen d subpol a novel sucrose-b...   \n",
       "952526  active-site mutants of the vanc2 d-alanyl-d-se...   \n",
       "952541  in vitro atp regeneration from polyphosphate a...   \n",
       "\n",
       "                                           chemical_names  \n",
       "150     [calcein-am, cyclosporin a, dexamethasone, dil...  \n",
       "271     [sodium dodecyl sulfate, geranyl pyrophosphate...  \n",
       "295                          [lysine, pyruvate, cysteine]  \n",
       "296     [pyruvate, cysteine, lysine, o-phthalaldehyde,...  \n",
       "298                         [pyruvate, oxaloacetate, adp]  \n",
       "...                                                   ...  \n",
       "952414                  [vancomycin, d-alanine-d-alanine]  \n",
       "952450                       [tetrahydrofurane, methanol]  \n",
       "952517                                    [sucrose, kato]  \n",
       "952526                   [d-alanyl-d-alanine, vancomycin]  \n",
       "952541                              [adenylate, amp, atp]  \n",
       "\n",
       "[44724 rows x 2 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re_check_several_chem_uncleaned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_6320/496917810.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  re_check_several_chem_uncleaned_df[\"text\"] = re_check_several_chem_uncleaned_df[[\"sentence\", \"chemical_names\"]].values.tolist()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['univ sydney, dept pharm, sydney, nsw 2006, australia abstract two distinct mg2+-atpase activities were isolated from triton x-100-solubilized human erythrocyte membranes using a combination of calmodulin-agarose to remove ca2+-atpase and ion exchange chromatography to separate the mg2+-atpase activities',\n",
       " ['ca2+', 'mg2+']]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re_check_several_chem_uncleaned_df[\"text\"] = re_check_several_chem_uncleaned_df[[\"sentence\", \"chemical_names\"]].values.tolist()\n",
    "re_check_several_chem_uncleaned_df.iloc[47][\"text\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n",
      "['6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst', ['gsh', '6-dhsg', 'gs-6-dhsg']]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "includes reaction component words: -1\n",
      "sep conversion words: True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for index in range(500, 520):\n",
    "    sentence = re_check_several_chem_uncleaned_df[\"text\"].iloc[index]\n",
    "    print(arguments)\n",
    "    print(\"general chemicals: \" + str(general_chemical(arguments)))\n",
    "    print(\"adjacent mentions: \" + str(adjacent_mentions(arguments)))\n",
    "    print(\"sep or: \" + str(sep_or(arguments)))\n",
    "    print(\"sep and: \" + str(sep_and(arguments)))\n",
    "    print(\"sep comma: \" + str(sep_comma(arguments)))\n",
    "    print(\"sep via: \" + str(sep_via(arguments)))\n",
    "    print(\"sep sym: \" + str(sep_sym(arguments)))\n",
    "    print(\"followed ase: \" + str(followed_ase(arguments)))\n",
    "    print(\"group: \" + str(group(arguments)))\n",
    "    print(\"followed by noun: \" + str(followed_by_noun(arguments)))\n",
    "    print(\"sep verb: \" + str(sep_verb(arguments)))\n",
    "    print(\"sep adverb: \" + str(sep_adverb(arguments)))\n",
    "    print(\"includes reaction words: \" + str(includes_reaction_words(arguments)))\n",
    "    print(\"includes reaction component words: \" + str(includes_reaction_component_words(arguments)))\n",
    "    print(\"sep conversion words: \" + str(sep_converstion_words(arguments)))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### snorkel code for model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.labeling import PandasLFApplier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:01<00:00, 61.73it/s]\n"
     ]
    }
   ],
   "source": [
    "# Define the set of labeling functions (LFs)\n",
    "lfs = [solution_words, mixture_words, physical_words, gene_words, structural_words, general_chemical, group,\n",
    "    includes_amino_acids, paper_artifacts, no_terms, adjacent_mentions, sep_or, sep_and, sep_comma, sep_via,\n",
    "    sep_sym, followed_ase, followed_by_noun, sep_verb, sep_adverb, includes_oxidation_words,\n",
    "    includes_combustion_words, includes_neutralization_words, includes_catalyze_words, includes_combination_words,\n",
    "    includes_decomposition_words, includes_replace_words, includes_reaction_words,\n",
    "    includes_reaction_component_words, comparison, sep_converstion_words]\n",
    "\n",
    "# Apply the LFs to the unlabeled training data\n",
    "df_train = re_check_several_chem_uncleaned_df[:100]\n",
    "applier = PandasLFApplier(lfs=lfs)\n",
    "L_train = applier.apply(df=df_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-1, -1,  0, ...,  1, -1,  1],\n",
       "       [-1, -1,  0, ..., -1, -1,  1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       ...,\n",
       "       [-1, -1,  0, ..., -1, -1,  1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1]])"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>solution_words</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mixture_words</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>physical_words</th>\n",
       "      <td>2</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_words</th>\n",
       "      <td>3</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structural_words</th>\n",
       "      <td>4</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>general_chemical</th>\n",
       "      <td>5</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group</th>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_amino_acids</th>\n",
       "      <td>7</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paper_artifacts</th>\n",
       "      <td>8</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_terms</th>\n",
       "      <td>9</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>adjacent_mentions</th>\n",
       "      <td>10</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_or</th>\n",
       "      <td>11</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_and</th>\n",
       "      <td>12</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_comma</th>\n",
       "      <td>13</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_via</th>\n",
       "      <td>14</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_sym</th>\n",
       "      <td>15</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>followed_ase</th>\n",
       "      <td>16</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>followed_by_noun</th>\n",
       "      <td>17</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_verb</th>\n",
       "      <td>18</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_adverb</th>\n",
       "      <td>19</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_oxidation_words</th>\n",
       "      <td>20</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combustion_words</th>\n",
       "      <td>21</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_neutralization_words</th>\n",
       "      <td>22</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_catalyze_words</th>\n",
       "      <td>23</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combination_words</th>\n",
       "      <td>24</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_decomposition_words</th>\n",
       "      <td>25</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_replace_words</th>\n",
       "      <td>26</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_words</th>\n",
       "      <td>27</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_component_words</th>\n",
       "      <td>28</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>comparison</th>\n",
       "      <td>29</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sep_converstion_words</th>\n",
       "      <td>30</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    j Polarity  Coverage  Overlaps  Conflicts\n",
       "solution_words                      0      [0]      0.01      0.01       0.01\n",
       "mixture_words                       1      [0]      0.01      0.01       0.01\n",
       "physical_words                      2      [0]      0.27      0.27       0.24\n",
       "gene_words                          3      [0]      0.26      0.26       0.22\n",
       "structural_words                    4      [0]      0.37      0.36       0.30\n",
       "general_chemical                    5      [0]      0.56      0.56       0.49\n",
       "group                               6      [0]      0.21      0.21       0.19\n",
       "includes_amino_acids                7      [0]      0.11      0.11       0.11\n",
       "paper_artifacts                     8      [0]      0.25      0.25       0.21\n",
       "no_terms                            9      [0]      0.21      0.21       0.21\n",
       "adjacent_mentions                  10      [0]      0.04      0.04       0.03\n",
       "sep_or                             11      [0]      0.02      0.02       0.02\n",
       "sep_and                            12      [0]      0.09      0.09       0.09\n",
       "sep_comma                          13      [0]      0.05      0.05       0.05\n",
       "sep_via                            14       []      0.00      0.00       0.00\n",
       "sep_sym                            15      [0]      0.18      0.16       0.11\n",
       "followed_ase                       16      [0]      0.48      0.48       0.42\n",
       "followed_by_noun                   17      [0]      0.60      0.60       0.53\n",
       "sep_verb                           18      [1]      0.52      0.52       0.52\n",
       "sep_adverb                         19      [0]      0.20      0.20       0.19\n",
       "includes_oxidation_words           20      [1]      0.08      0.08       0.08\n",
       "includes_combustion_words          21      [1]      0.08      0.08       0.08\n",
       "includes_neutralization_words      22       []      0.00      0.00       0.00\n",
       "includes_catalyze_words            23      [1]      0.38      0.38       0.38\n",
       "includes_combination_words         24      [1]      0.02      0.02       0.02\n",
       "includes_decomposition_words       25       []      0.00      0.00       0.00\n",
       "includes_replace_words             26       []      0.00      0.00       0.00\n",
       "includes_reaction_words            27      [1]      0.49      0.49       0.49\n",
       "includes_reaction_component_words  28      [1]      0.23      0.23       0.23\n",
       "comparison                         29      [0]      0.09      0.09       0.06\n",
       "sep_converstion_words              30      [1]      0.24      0.24       0.24"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from snorkel.labeling import LFAnalysis\n",
    "\n",
    "LFAnalysis(L=L_train, lfs=lfs).lf_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.labeling.model import MajorityLabelVoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.labeling.model import LabelModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_6320/428493704.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_train[\"label\"] = label_model.predict(L=L_train, tie_break_policy=\"abstain\")\n"
     ]
    }
   ],
   "source": [
    "# Train the label model and compute the training labels\n",
    "label_model = LabelModel(cardinality=2, verbose=True)\n",
    "label_model.fit(L_train, n_epochs=500, log_freq=50, seed=123)\n",
    "df_train[\"label\"] = label_model.predict(L=L_train, tie_break_policy=\"abstain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    79\n",
       "1    21\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train[\"label\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train.to_csv(r'../labeled.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e1fffad673803e23a98281b7492a34c559235864658941242758d49259d94b3b"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit (windows store)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
