{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "## iGEM labeling functions"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## example"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "source": [
    "import pandas as pd\r\n",
    "import numpy as np\r\n",
    "import os\r\n",
    "import nltk\r\n",
    "import regex as re\r\n",
    "import ast\r\n",
    "from collections import defaultdict\r\n",
    "# from snorkel.labeling import labeling_function"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "source": [
    "# @labeling_function()\r\n",
    "def oxidation_with_sub_prod(sentence):\r\n",
    "    for word in oxidationWords:\r\n",
    "        if word in sentence.lower:\r\n",
    "            return TRUE\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### test data written by hand"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "source": [
    "test_data = [\r\n",
    "    [[\"carbon\", \"oxygen\"], \"carbon was oxidized by the oxygen\"],\r\n",
    "    [[\"carbon\", \"oxygen\", \"amino acid\", \"cassie\"], \"the cassie amino acid is connected to the carbon which was oxidixed by the oxygen inside the dna\"],\r\n",
    "    \r\n",
    "    [[\"L-GalDH\"], \"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\"],\r\n",
    "    [[\"L-GalDH\"], \"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\"],\r\n",
    "    [[\"L-GalL dehydrogenase\"], \"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\"],\r\n",
    "    [[\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"], \"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\"]\r\n",
    "]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### setting up abstain"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "source": [
    "ABSTAIN = -1"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### labeling functions + small tests written"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "source": [
    "# LF_general_chemical\r\n",
    "# If one of the chemicals is in a list of common chemical terms such as amino acid, sugar, adenosine, amide, adenine, etc..., we label FALSE\r\n",
    "# https://www.hach.com/chemGlossary   \r\n",
    "common_chemical_terms = [\"amino acid\", \"sugar\", \"adenosine\", \"amide\", \"adenine\", \"acid\", \"algaecide\", \"amines\", \"base\", \"biocides\",\r\n",
    "    \"buffer\", \"clarifier\", \"diluent\", \"hydrocarbon\", \"hydrogenation\", \"molecule\", \"nutrients\", \"polymer\", \"peptide\", \"polypeptide\", \"solute\",\r\n",
    "    \"solvent\", \"suspended\", \"mixture\", \"saturated\", \"unsaturated\"]\r\n",
    "def general_chemical(chemicals):\r\n",
    "    for chem in chemicals:\r\n",
    "        if (chem in common_chemical_terms):\r\n",
    "            return (False)\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "source": [
    "# LF_adjacent_mentions\r\n",
    "# If the chemicals are adjacent, we label FALSE\r\n",
    "def adjacent_mentions(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        for index_2 in range(len(chemicals)):\r\n",
    "            if ((chemicals[index] + \" \" + chemicals[index_2]) in sentence):\r\n",
    "                return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "source": [
    "# LF_gene_words\r\n",
    "# If one of the chemicals is in a list of genetic terms such as reductase, dna, gene, we label FALSE\r\n",
    "# \r\n",
    "common_genetic_terms = [\"reductase\", \"dna\", \"gene\", \"allele\", \"locus\", \"genotype\", \"phenotype\", \"dominant\", \"recessive\", \"additive\", \"phenoset\",\r\n",
    "    \"diallelic\", \"multiallelic\", \"polyallelic\", \"monomorphic\", \"monoallelic\", \"polymorphism\", \"mutation\", \"complex\", \"trait\", \"multifactorial\",\r\n",
    "    \"polygenic\", \"monogenic\", \"mixed model\", \"transmission probability\", \"transition probability\", \"epistasis\", \"interaction\", \"pleiotropy\",\r\n",
    "    \"quantitative trait locus\", \"probit\", \"logit\", \"penetrance\", \"transformation\", \"scale of measurement\", \"identity by descent\", \"identity in state\",\r\n",
    "    \"haplotype\", \"phase\", \"multilocus\", \"genotype\", \"allelic\", \"association\", \"linkage\", \"disequilibrium\", \"gametic\"]\r\n",
    "def gene_words(chemicals):\r\n",
    "    for chem in chemicals:\r\n",
    "        if (chem in common_genetic_terms):\r\n",
    "            return False\r\n",
    "    return ABSTAIN\r\n",
    "    "
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "source": [
    "# LF_sep_or\r\n",
    "# If the chemicals are separated by or, we label FALSE\r\n",
    "def sep_or(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        for index_2 in range(len(chemicals)):\r\n",
    "            if (chemicals[index] + \" or \" + chemicals[index_2] in sentence):\r\n",
    "                return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "source": [
    "# LF_sep_and\r\n",
    "# If the chemicals are separated by and, we label FALSE\r\n",
    "def sep_and(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        for index_2 in range(len(chemicals)):\r\n",
    "            if (chemicals[index] + \" and \" + chemicals[index_2] in sentence):\r\n",
    "                return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "source": [
    "# LF_sep_comma\r\n",
    "# If the chemicals are separated by a comma, we label FALSE\r\n",
    "def sep_comma(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        for index_2 in range(len(chemicals)):\r\n",
    "            if (chemicals[index] + \", \" + chemicals[index_2] in sentence):\r\n",
    "                return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "source": [
    "# LF_sep_via\r\n",
    "# If the chemicals are separated by via, in, etc, we label FALSE\r\n",
    "common_via_words = [\"via\", \"in\"]\r\n",
    "def sep_via(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        for index_2 in range(len(chemicals)):\r\n",
    "            for term in common_via_words:\r\n",
    "                if (chemicals[index] + \" \" + term + \" \" + chemicals[index_2] in sentence):\r\n",
    "                    return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "source": [
    "# LF_sep_sym\r\n",
    "# If the chemicals are separated by a single character, we label FALSE  \r\n",
    "def sep_sym(chemicals, sentence):\r\n",
    "    for index in range(len(chemicals)):\r\n",
    "        chem_1_index = sentence.find(chemicals[index])\r\n",
    "        chem_1_len = len(chemicals[index])\r\n",
    "        while (0 <= chem_1_index < len(sentence)):\r\n",
    "            for index_2 in range(len(chemicals)):\r\n",
    "                chem_2_index = sentence.find(chemicals[index_2])\r\n",
    "                while (0 <= chem_2_index < len(sentence)):\r\n",
    "                    if (sep_sym_helper(chem_1_index, chem_1_len, chem_2_index)):\r\n",
    "                        return False\r\n",
    "                    chem_2_index = sentence.find(chemicals[index_2], chem_2_index + 1)\r\n",
    "            chem_1_index = sentence.find(chemicals[index], chem_1_index + 1)\r\n",
    "    return ABSTAIN\r\n",
    "\r\n",
    "def sep_sym_helper(index_1, length, index_2):\r\n",
    "    if (index_1 + length + 1== index_2):\r\n",
    "        return True\r\n",
    "    return False"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "source": [
    "# testing separated by a single character (ensuring it works with repeated chemicals)\r\n",
    "sep_sym([\"cassie\", \"cas\"], \"cassie is here and her name cassie cas is cas cassie\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "metadata": {},
     "execution_count": 305
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "source": [
    "# LF_followed_ase\r\n",
    "# If one of the chemicals is followed by a word that ends with -ase, we label FALSE\r\n",
    "def followed_ase(chemicals, sentence):\r\n",
    "    sentence = sentence.replace('.', '')\r\n",
    "    sentence = sentence.replace(',', '')\r\n",
    "    words = sentence.split(\" \")\r\n",
    "    for chem in chemicals:\r\n",
    "        index = -1\r\n",
    "        while (index < len(words)):\r\n",
    "            if not(chem in words[index + 1:]):\r\n",
    "                index = len(words)\r\n",
    "                break\r\n",
    "            if (\" \" in chem):\r\n",
    "                chem_split = chem.split(\" \")\r\n",
    "                index = words.index(chem_split[-1], index + 1)\r\n",
    "            else:\r\n",
    "                index = words.index(chem, index + 1)\r\n",
    "            if (0 < index + 1 < len(words) and words[index + 1][-3:] == \"ase\"):\r\n",
    "                return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "source": [
    "# test followed_ase (making sure it works if the chem is repeated)\r\n",
    "followed_ase([\"carbon\", \"oxygen\"], \"carbon fiber and carbon lactase with oxygen\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "metadata": {},
     "execution_count": 307
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "source": [
    "# LF_group\r\n",
    "# If there is a close mention of a functional chemical group, we label FALSE (as it is more likely descriptive of a structure than of a reaction)\r\n",
    "# https://www.masterorganicchemistry.com/2010/10/06/functional-groups-organic-chemistry/\r\n",
    "common_functional_groups = [\"alkane\", \"alkene\", \"alkyne\", \"benzene ring\", \"phenyl\", \"amine\", \"alcohol\", \"ether\", \"alkyl halide\", \"thiol\",\r\n",
    "    \"aldehyde\", \"ketone\", \"ester\", \"carboxylic acid\", \"amide\", \"nitrile\", \"epoxide\", \"disulfide\", \"imine\", \"acid chloride\", \"anhydride\", \"nitro\",\r\n",
    "    \"sulfide\", \"thioether\"]\r\n",
    "def group(sentence):\r\n",
    "    for group in common_functional_groups:\r\n",
    "        if (group in sentence):\r\n",
    "            return False\r\n",
    "    return ABSTAIN\r\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "source": [
    "# LF_followed_by_noun\r\n",
    "# If one of the chemicals is followed by a noun, we label FALSE\r\n",
    "def followed_by_noun(chemicals, sentence):\r\n",
    "    sentence = sentence.replace(',', '')\r\n",
    "    words = sentence.split(\" \")\r\n",
    "    tagged = nltk.pos_tag(words)\r\n",
    "    for chem in chemicals:\r\n",
    "        index = -1\r\n",
    "        while (index < len(words)):\r\n",
    "            if not(chem in words[index + 1:]):\r\n",
    "                index = len(words)\r\n",
    "                break\r\n",
    "            if (\" \" in chem):\r\n",
    "                chem_split = chem.split(\" \")\r\n",
    "                index = words.index(chem_split[-1], index + 1)\r\n",
    "            else:\r\n",
    "                index = words.index(chem, index + 1)\r\n",
    "            if (index + 1 < len(words)):\r\n",
    "                if (tagged[index + 1][1] == \"NN\"):\r\n",
    "                    return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "source": [
    "# testing followed_by_noun (ensuring it works with repeats)\r\n",
    "followed_by_noun([\"cassie\", \"apple\"], \"cassie eating apple, eating for cassie table\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "metadata": {},
     "execution_count": 310
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "source": [
    "# test cases\r\n",
    "print(followed_by_noun([\"L-GalDH\"], \"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\"))\r\n",
    "print(followed_by_noun([\"L-GalDH\"], \"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\"))\r\n",
    "print(followed_by_noun([\"L-GalL dehydrogenase\"], \"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\"))\r\n",
    "print(followed_by_noun([\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"], \"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\"))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "False\n",
      "False\n",
      "-1\n",
      "False\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "source": [
    "# LF_sep_verb\r\n",
    "# If the chemicals are separated by a verb, we label TRUE\r\n",
    "def sep_verb(chemicals, sentence):\r\n",
    "    sentence = sentence.replace(',', '')\r\n",
    "    words = sentence.split(\" \")\r\n",
    "    indexes = []\r\n",
    "    for chem in chemicals:\r\n",
    "        index = -1\r\n",
    "        while (index < len(words)):\r\n",
    "            if (\" \" in chem):\r\n",
    "                chem_split = chem.split(\" \")\r\n",
    "                if not(chem_split[-1] in words[index + 1:]):\r\n",
    "                    index = len(words)\r\n",
    "                    break\r\n",
    "                index = words.index(chem_split[-1], index + 1)\r\n",
    "            else:\r\n",
    "                if not(chem in words[index + 1:]):\r\n",
    "                    index = len(words)\r\n",
    "                    break\r\n",
    "                index = words.index(chem, index + 1)\r\n",
    "            indexes.append(index)\r\n",
    "    min_index = min(indexes)\r\n",
    "    max_index = max(indexes)\r\n",
    "    tagged = nltk.pos_tag(words)\r\n",
    "    for index in range(min_index + 1, max_index):\r\n",
    "        if (tagged[index][1][:2] == \"VB\"):\r\n",
    "            return True\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "source": [
    "# testing sep_verb (works when there are several instances)\r\n",
    "sep_verb([\"cassie\", \"cas\"], \"cassie cas cassie hi there oxidized cas\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "execution_count": 313
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "source": [
    "# test cases\r\n",
    "print(sep_verb([\"L-GalDH\"], \"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\"))\r\n",
    "print(sep_verb([\"L-GalDH\"], \"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\"))\r\n",
    "print(sep_verb([\"L-GalL dehydrogenase\"], \"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\"))\r\n",
    "print(sep_verb([\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"], \"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\"))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "-1\n",
      "-1\n",
      "-1\n",
      "True\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "source": [
    "# LF_sep_adverb\r\n",
    "# If the chemicals are separated by a adverb, we label FALSE\r\n",
    "def sep_adverb(chemicals, sentence):\r\n",
    "    sentence = sentence.replace(',', '')\r\n",
    "    words = sentence.split(\" \")\r\n",
    "    indexes = []\r\n",
    "    for chem in chemicals:\r\n",
    "        index = -1\r\n",
    "        while (index < len(words)):\r\n",
    "            if (\" \" in chem):\r\n",
    "                chem_split = chem.split(\" \")\r\n",
    "                if not(chem_split[-1] in words[index + 1:]):\r\n",
    "                    index = len(words)\r\n",
    "                    break\r\n",
    "                index = words.index(chem_split[-1], index + 1)\r\n",
    "            else:\r\n",
    "                if not(chem in words[index + 1:]):\r\n",
    "                    index = len(words)\r\n",
    "                    break\r\n",
    "                index = words.index(chem, index + 1)\r\n",
    "            indexes.append(index)\r\n",
    "    min_index = min(indexes)\r\n",
    "    max_index = max(indexes)\r\n",
    "    tagged = nltk.pos_tag(words)\r\n",
    "    for index in range(min_index + 1, max_index):\r\n",
    "        if (tagged[index][1][:2] == \"RB\"):\r\n",
    "            return False\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "source": [
    "# testing sep_adverb (for repeated chemicals)\r\n",
    "sep_adverb([\"cassie\", \"cas\"], \"cassie happily danced to the moon with cas\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "metadata": {},
     "execution_count": 316
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "source": [
    "# test cases\r\n",
    "print(sep_adverb([\"L-GalDH\"], \"The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.\"))\r\n",
    "print(sep_adverb([\"L-GalDH\"], \"Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.\"))\r\n",
    "print(sep_adverb([\"L-GalL dehydrogenase\"], \"We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).\"))\r\n",
    "print(sep_adverb([\"L-GalDH\", \"dehydro-AsA\", \"L-GalL\"], \"The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.\"))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "-1\n",
      "-1\n",
      "-1\n",
      "False\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "source": [
    "# LF_includes_reaction_words\r\n",
    "# If the sentence contains reactions words, we label True\r\n",
    "reaction_words = [\"convert\", \"yeild\", \"produce\", \"make\", \"react\", \"create\", \"synthesize\", \"conversion\",\r\n",
    "    \"transformation\", \"transform\", \"synthesise\", \"burn\", \"rust\", \"ferment\", \"explode\", \"agent\", \"catalyze\"\r\n",
    "    \"combust\", \"corrode\", \"decompose\", \"oxidize\", \"neutralize\", \"oxidization\", \"neutralization\", \"combustion\",\r\n",
    "    \"redox\", \"methylaion\", \"gas-forming\", \"displacement\", \"combination\", \"exchange\", \"precipitate\",\r\n",
    "    \"precipitation\", \"transfer\", \"through\", \"by\"]\r\n",
    "def includes_reaction_words(sentence):\r\n",
    "    for word in reaction_words:\r\n",
    "        if(word in sentence):\r\n",
    "            return True\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "source": [
    "# LF_sep_conversion_words\r\n",
    "# If the sentence contains to, from, into, etc., we label True\r\n",
    "conversion_words = [\"to\", \"from\", \"into\", \"becomes\", \"became\"]\r\n",
    "def sep_converstion_words(chemicals, sentence):\r\n",
    "    for word in conversion_words:\r\n",
    "        for index in range(len(chemicals)):\r\n",
    "            for index_2 in range(len(chemicals)):\r\n",
    "                if (chemicals[index] + \" \" + word + \" \" + chemicals[index_2] in sentence):\r\n",
    "                    return True\r\n",
    "    return ABSTAIN"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "source": [
    "# test sep_conversion_words\r\n",
    "sep_converstion_words([\"cassie\", \"cas\"], \"cassie into cas\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "execution_count": 355
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### my todo + notes"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "source": [
    "# things to fix\r\n",
    "    # create specific reation verb labeling (complete)\r\n",
    "        # oxidation\r\n",
    "        # reduction\r\n",
    "        # combustion\r\n",
    "        # composition\r\n",
    "        # decomposition\r\n",
    "        # etc\r\n",
    "    # creating more test cases\r\n",
    "    # current test sentences from\r\n",
    "        # jacob's csv\r\n",
    "        # https://academic.oup.com/pcp/article/45/9/1271/1857717\r\n",
    "    # ways to clean data\r\n",
    "        # ensure there are at least two chemicals (complete)\r\n",
    "        # get rid of the greek alphabet (complete)\r\n",
    "        # get rid of periods and () and ; (complete)\r\n",
    "        # get rid of repeat chems (complete)\r\n",
    "    # need to figure out how to deal with things in the sentence twice (complete)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### testing on data written in this file"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "source": [
    "for data in test_data:\r\n",
    "    print(data)\r\n",
    "    print(\"general chemicals\")\r\n",
    "    print(general_chemical(data[0]))\r\n",
    "    print(\"adjacent mentions\")\r\n",
    "    print(adjacent_mentions(data[0], data[1]))\r\n",
    "    print(\"sep or\")\r\n",
    "    print(sep_or(data[0], data[1]))\r\n",
    "    print(\"sep and\")\r\n",
    "    print(sep_and(data[0], data[1]))\r\n",
    "    print(\"sep comma\")\r\n",
    "    print(sep_comma(data[0], data[1]))\r\n",
    "    print(\"sep via\")\r\n",
    "    print(sep_via(data[0], data[1]))\r\n",
    "    print(\"sep sym\")\r\n",
    "    print(sep_sym(data[0], data[1]))\r\n",
    "    print(\"followed ase\")\r\n",
    "    print(followed_ase(data[0], data[1]))\r\n",
    "    print(\"group\")\r\n",
    "    print(group(data[1]))\r\n",
    "    print(\"followed by noun\")\r\n",
    "    print(followed_by_noun(data[0], data[1]))\r\n",
    "    print(\"sep verb\")\r\n",
    "    print(sep_verb(data[0], data[1]))\r\n",
    "    print(\"sep adverb\")\r\n",
    "    print(sep_adverb(data[0], data[1]))\r\n",
    "    print()"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[['carbon', 'oxygen'], 'carbon was oxidized by the oxygen']\n",
      "general chemicals\n",
      "-1\n",
      "adjacent mentions\n",
      "-1\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "-1\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "-1\n",
      "sep verb\n",
      "True\n",
      "sep adverb\n",
      "-1\n",
      "\n",
      "[['carbon', 'oxygen', 'amino acid', 'cassie'], 'the cassie amino acid is connected to the carbon which was oxidixed by the oxygen inside the dna']\n",
      "general chemicals\n",
      "False\n",
      "adjacent mentions\n",
      "False\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "False\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "False\n",
      "sep verb\n",
      "True\n",
      "sep adverb\n",
      "-1\n",
      "\n",
      "[['L-GalDH'], 'The active fractions from the gel filtration column were further separated by native PAGE, followed by staining for L-GalDH activity.']\n",
      "general chemicals\n",
      "-1\n",
      "adjacent mentions\n",
      "-1\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "-1\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "False\n",
      "sep verb\n",
      "-1\n",
      "sep adverb\n",
      "-1\n",
      "\n",
      "[['L-GalDH'], 'Southern blot analysis revealed that the spinach L-GalDH gene occurs in a single copy.']\n",
      "general chemicals\n",
      "-1\n",
      "adjacent mentions\n",
      "-1\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "-1\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "False\n",
      "sep verb\n",
      "-1\n",
      "sep adverb\n",
      "-1\n",
      "\n",
      "[['L-GalL dehydrogenase'], 'We reported that transcripts of L-GalL dehydrogenase (L-GalLDH), which functions at the last step of the biosynthetic pathway, were expressed in leaves, stems and roots at almost the same level (Yabuta et al. 2000).']\n",
      "general chemicals\n",
      "-1\n",
      "adjacent mentions\n",
      "-1\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "-1\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "-1\n",
      "sep verb\n",
      "-1\n",
      "sep adverb\n",
      "-1\n",
      "\n",
      "[['L-GalDH', 'dehydro-AsA', 'L-GalL'], 'The L-GalDH activity was not affected by the addition of 1 mM dehydro-AsA (DHA) or 1 mM L-GalL.']\n",
      "general chemicals\n",
      "-1\n",
      "adjacent mentions\n",
      "-1\n",
      "sep or\n",
      "-1\n",
      "sep and\n",
      "-1\n",
      "sep comma\n",
      "-1\n",
      "sep via\n",
      "-1\n",
      "sep sym\n",
      "-1\n",
      "followed ase\n",
      "-1\n",
      "group\n",
      "-1\n",
      "followed by noun\n",
      "False\n",
      "sep verb\n",
      "True\n",
      "sep adverb\n",
      "False\n",
      "\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### importing data in and cleaning it"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "source": [
    "sentence_df = pd.read_csv(\"sentence_data_cleaned_jacob.csv\")\r\n",
    "sentence_with_chem_df = sentence_df.dropna()\r\n",
    "several_chem_df = sentence_with_chem_df.loc[sentence_with_chem_df[\"chemicals\"].str.contains(\",\")]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "source": [
    "sentence_chem_df = several_chem_df[[\"sentence\", \"chemicals\"]]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "source": [
    "greek_alphabet = {\r\n",
    "    u'\\u0393': 'Gamma',\r\n",
    "    u'\\u0394': 'Delta',\r\n",
    "    u'\\u0398': 'Theta',\r\n",
    "    u'\\u039B': 'Lamda',\r\n",
    "    u'\\u039E': 'Xi',\r\n",
    "    u'\\u03A0': 'Pi',\r\n",
    "    u'\\u03A3': 'Sigma',\r\n",
    "    u'\\u03A6': 'Phi',\r\n",
    "    u'\\u03A7': 'Chi',\r\n",
    "    u'\\u03A8': 'Psi',\r\n",
    "    u'\\u03A9': 'Omega',\r\n",
    "    u'\\u03B1': 'alpha',\r\n",
    "    u'\\u03B2': 'beta',\r\n",
    "    u'\\u03B3': 'gamma',\r\n",
    "    u'\\u03B4': 'delta',\r\n",
    "    u'\\u03B5': 'epsilon',\r\n",
    "    u'\\u03B6': 'zeta',\r\n",
    "    u'\\u03B7': 'eta',\r\n",
    "    u'\\u03B8': 'theta',\r\n",
    "    u'\\u03B9': 'iota',\r\n",
    "    u'\\u03BA': 'kappa',\r\n",
    "    u'\\u03BB': 'lamda',\r\n",
    "    u'\\u03BC': 'mu',\r\n",
    "    u'\\u03BD': 'nu',\r\n",
    "    u'\\u03BE': 'xi',\r\n",
    "    u'\\u03BF': 'omicron',\r\n",
    "    u'\\u03C0': 'pi',\r\n",
    "    u'\\u03C1': 'rho',\r\n",
    "    u'\\u03C3': 'sigma',\r\n",
    "    u'\\u03C4': 'tau',\r\n",
    "    u'\\u03C5': 'upsilon',\r\n",
    "    u'\\u03C6': 'phi',\r\n",
    "    u'\\u03C7': 'chi',\r\n",
    "    u'\\u03C8': 'psi',\r\n",
    "    u'\\u03C9': 'omega'\r\n",
    "}"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "source": [
    "'β' in greek_alphabet.keys()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "execution_count": 323
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "source": [
    "def remove_greek(words):\r\n",
    "    for letter in greek_alphabet.keys():\r\n",
    "        words = words.replace(letter, greek_alphabet[letter])\r\n",
    "    return words\r\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "source": [
    "# greek test\r\n",
    "remove_greek(\"11β-hydroxysteroid\")"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'11beta-hydroxysteroid'"
      ]
     },
     "metadata": {},
     "execution_count": 325
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "source": [
    "def chem_into_array(chemicals):\r\n",
    "    chemicals = remove_greek(chemicals)\r\n",
    "    chemicals = chemicals.lower()\r\n",
    "    sample_chems = chemicals.replace(\",\", \"\")\r\n",
    "    sample_chems_list = sample_chems.split(\" \")\r\n",
    "    sample_chems_list = list(set(sample_chems_list))\r\n",
    "    return sample_chems_list"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "source": [
    "def sentence_cleaned(sentence):\r\n",
    "    sentence = remove_greek(sentence)\r\n",
    "    sentence = sentence.lower()\r\n",
    "    sentence = sentence.replace(\".\", \"\")\r\n",
    "    sentence = sentence.replace(\")\", \"\")\r\n",
    "    sentence = sentence.replace(\"(\", \"\")\r\n",
    "    sentence = sentence.replace(\";\", \"\")\r\n",
    "    sentence = sentence.replace(\":\", \"\")\r\n",
    "    return sentence"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "source": [
    "sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\r\n",
    "sentence_chem_df"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_23384/942711163.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>background and aims glutamic acid decarboxylas...</td>\n",
       "      <td>glutamate, GABA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>interestingly, the allele of pro1 was shown to...</td>\n",
       "      <td>gamma-glutamyl, gamma-glutamyl, L-proline, L-g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>we concluded that fdh has no direct role in th...</td>\n",
       "      <td>Ser, formate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>even though the activities of mat and gnmt wer...</td>\n",
       "      <td>S-adenosylmethionine, S-adenosylhomocysteine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>one of the enzymes responsible for the product...</td>\n",
       "      <td>KA, kynurenine, glutamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>purpose the fluoropyrimidine carbamate capecit...</td>\n",
       "      <td>capecitabine, 5-fluorouracil, 5-FU, thymidine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>carnitine acetyltransferases crat catalyze the...</td>\n",
       "      <td>Carnitine, acetyl-CoA, carnitine, acetylcarnitine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>l-serine dehydratase sdh, a member of the beta...</td>\n",
       "      <td>L-serine, L-serine, L-threonine, pyruvate, 2-o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>spermidine/spermine n1-acetyltransferase ssat ...</td>\n",
       "      <td>polyamine, spermidine, spermine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>astrocytes may play a role in these manifestat...</td>\n",
       "      <td>glutamate, glutamine, glutamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>reduced synthesis of nitric oxide no contribut...</td>\n",
       "      <td>NO, L-arginine, nitric-oxide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>Estrogens, androgens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>background histamine synthesized by histidine ...</td>\n",
       "      <td>Histamine, histidine, L-histidine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>methionine synthase reductase mtrr is an enzym...</td>\n",
       "      <td>Methionine, Hcy, methionine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>they all expressed ass, but not ornithine tran...</td>\n",
       "      <td>ornithine, ornithine, arginine, citrulline, ar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>unlike mouse retsat mretsat, zretsat a had an ...</td>\n",
       "      <td>all-trans-retinol, all-trans-13,14-dihydroreti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>ldh is responsible for pyruvate conversion to ...</td>\n",
       "      <td>pyruvate, lactate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>objective the aim was to test whether the comm...</td>\n",
       "      <td>PUFA, ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>in kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>cAMP, cAMP, AMP, AMP, adenosine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>adenylosuccinate synthetase adss catalyzes the...</td>\n",
       "      <td>Adenylosuccinate, Mg2+, aspartate, adenylosucc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>risperidone is metabolized to its active metab...</td>\n",
       "      <td>Risperidone, 9-hydroxyrisperidone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>aldo-keto reductases akrs metabolize a wide ra...</td>\n",
       "      <td>Aldo-keto, PAHs, o-quinones, oxygen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>aldose reductase ar catalyzes the reduction of...</td>\n",
       "      <td>Aldose, aldehydes, alcohol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>11β-hydroxysteroid, cortisol, cortisone, glucose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>esters, amides, thioesters, alcohols, amines, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>pip5k1b encodes phosphatidylinositol 4-phospha...</td>\n",
       "      <td>phosphatidylinositol-4,5-bisphosphate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>6-dhsg was metabolised by gsh to form a gsh co...</td>\n",
       "      <td>6-DHSG, GSH, GSH, GS-6-DHSG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence  \\\n",
       "197   background and aims glutamic acid decarboxylas...   \n",
       "252   interestingly, the allele of pro1 was shown to...   \n",
       "282   we concluded that fdh has no direct role in th...   \n",
       "318   even though the activities of mat and gnmt wer...   \n",
       "323   one of the enzymes responsible for the product...   \n",
       "439   purpose the fluoropyrimidine carbamate capecit...   \n",
       "465   carnitine acetyltransferases crat catalyze the...   \n",
       "484   l-serine dehydratase sdh, a member of the beta...   \n",
       "545   spermidine/spermine n1-acetyltransferase ssat ...   \n",
       "561   astrocytes may play a role in these manifestat...   \n",
       "607   reduced synthesis of nitric oxide no contribut...   \n",
       "661   estrogens are biosynthesised from androgens by...   \n",
       "682   background histamine synthesized by histidine ...   \n",
       "694   methionine synthase reductase mtrr is an enzym...   \n",
       "729   they all expressed ass, but not ornithine tran...   \n",
       "742   unlike mouse retsat mretsat, zretsat a had an ...   \n",
       "759   ldh is responsible for pyruvate conversion to ...   \n",
       "777   objective the aim was to test whether the comm...   \n",
       "789   in kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   adenylosuccinate synthetase adss catalyzes the...   \n",
       "905   risperidone is metabolized to its active metab...   \n",
       "995   aldo-keto reductases akrs metabolize a wide ra...   \n",
       "1086  aldose reductase ar catalyzes the reduction of...   \n",
       "1179  type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  carboxylesterases hydrolyze esters, amides, an...   \n",
       "1367  pip5k1b encodes phosphatidylinositol 4-phospha...   \n",
       "1383  6-dhsg was metabolised by gsh to form a gsh co...   \n",
       "\n",
       "                                              chemicals  \n",
       "197                                     glutamate, GABA  \n",
       "252   gamma-glutamyl, gamma-glutamyl, L-proline, L-g...  \n",
       "282                                        Ser, formate  \n",
       "318        S-adenosylmethionine, S-adenosylhomocysteine  \n",
       "323                           KA, kynurenine, glutamine  \n",
       "439       capecitabine, 5-fluorouracil, 5-FU, thymidine  \n",
       "465   Carnitine, acetyl-CoA, carnitine, acetylcarnitine  \n",
       "484   L-serine, L-serine, L-threonine, pyruvate, 2-o...  \n",
       "545                     polyamine, spermidine, spermine  \n",
       "561                     glutamate, glutamine, glutamine  \n",
       "607                        NO, L-arginine, nitric-oxide  \n",
       "661                                Estrogens, androgens  \n",
       "682                   Histamine, histidine, L-histidine  \n",
       "694                         Methionine, Hcy, methionine  \n",
       "729   ornithine, ornithine, arginine, citrulline, ar...  \n",
       "742   all-trans-retinol, all-trans-13,14-dihydroreti...  \n",
       "759                                   pyruvate, lactate  \n",
       "777                                           PUFA, ALA  \n",
       "789                     cAMP, cAMP, AMP, AMP, adenosine  \n",
       "881   Adenylosuccinate, Mg2+, aspartate, adenylosucc...  \n",
       "905                   Risperidone, 9-hydroxyrisperidone  \n",
       "995                 Aldo-keto, PAHs, o-quinones, oxygen  \n",
       "1086                         Aldose, aldehydes, alcohol  \n",
       "1179   11β-hydroxysteroid, cortisol, cortisone, glucose  \n",
       "1195  esters, amides, thioesters, alcohols, amines, ...  \n",
       "1367              phosphatidylinositol-4,5-bisphosphate  \n",
       "1383                        6-DHSG, GSH, GSH, GS-6-DHSG  "
      ]
     },
     "metadata": {},
     "execution_count": 328
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "source": [
    "sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)\r\n",
    "sentence_chem_df"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_23384/2741015264.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>background and aims glutamic acid decarboxylas...</td>\n",
       "      <td>[glutamate, gaba]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>interestingly, the allele of pro1 was shown to...</td>\n",
       "      <td>[l-glutamate, gamma-glutamyl, l-proline]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>we concluded that fdh has no direct role in th...</td>\n",
       "      <td>[formate, ser]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>even though the activities of mat and gnmt wer...</td>\n",
       "      <td>[s-adenosylmethionine, s-adenosylhomocysteine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>one of the enzymes responsible for the product...</td>\n",
       "      <td>[glutamine, kynurenine, ka]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>purpose the fluoropyrimidine carbamate capecit...</td>\n",
       "      <td>[capecitabine, thymidine, 5-fu, 5-fluorouracil]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>carnitine acetyltransferases crat catalyze the...</td>\n",
       "      <td>[carnitine, acetyl-coa, acetylcarnitine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>l-serine dehydratase sdh, a member of the beta...</td>\n",
       "      <td>[l-serine, 2-oxobutyrate, pyruvate, l-threonine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>spermidine/spermine n1-acetyltransferase ssat ...</td>\n",
       "      <td>[spermidine, spermine, polyamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>astrocytes may play a role in these manifestat...</td>\n",
       "      <td>[glutamate, glutamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>reduced synthesis of nitric oxide no contribut...</td>\n",
       "      <td>[l-arginine, no, nitric-oxide]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>estrogens are biosynthesised from androgens by...</td>\n",
       "      <td>[androgens, estrogens]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>background histamine synthesized by histidine ...</td>\n",
       "      <td>[histidine, l-histidine, histamine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>methionine synthase reductase mtrr is an enzym...</td>\n",
       "      <td>[methionine, hcy]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>they all expressed ass, but not ornithine tran...</td>\n",
       "      <td>[citrulline, arginine, ornithine]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>unlike mouse retsat mretsat, zretsat a had an ...</td>\n",
       "      <td>[all-trans-1314-dihydroretinol, all-trans-reti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>ldh is responsible for pyruvate conversion to ...</td>\n",
       "      <td>[pyruvate, lactate]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>objective the aim was to test whether the comm...</td>\n",
       "      <td>[ala, pufa]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>in kidneys, stimulation of adenylyl cyclase ca...</td>\n",
       "      <td>[amp, adenosine, camp]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>adenylosuccinate synthetase adss catalyzes the...</td>\n",
       "      <td>[gtp, gdp, mg2+, adenylosuccinate, aspartate]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>risperidone is metabolized to its active metab...</td>\n",
       "      <td>[9-hydroxyrisperidone, risperidone]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>aldo-keto reductases akrs metabolize a wide ra...</td>\n",
       "      <td>[o-quinones, aldo-keto, oxygen, pahs]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>aldose reductase ar catalyzes the reduction of...</td>\n",
       "      <td>[aldehydes, aldose, alcohol]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>type 2 11beta-hydroxysteroid dehydrogenase enc...</td>\n",
       "      <td>[cortisone, cortisol, 11beta-hydroxysteroid, g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>carboxylesterases hydrolyze esters, amides, an...</td>\n",
       "      <td>[amines, thiols, amides, alcohols, thioesters,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>pip5k1b encodes phosphatidylinositol 4-phospha...</td>\n",
       "      <td>[phosphatidylinositol-45-bisphosphate]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1383</th>\n",
       "      <td>6-dhsg was metabolised by gsh to form a gsh co...</td>\n",
       "      <td>[gs-6-dhsg, gsh, 6-dhsg]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence  \\\n",
       "197   background and aims glutamic acid decarboxylas...   \n",
       "252   interestingly, the allele of pro1 was shown to...   \n",
       "282   we concluded that fdh has no direct role in th...   \n",
       "318   even though the activities of mat and gnmt wer...   \n",
       "323   one of the enzymes responsible for the product...   \n",
       "439   purpose the fluoropyrimidine carbamate capecit...   \n",
       "465   carnitine acetyltransferases crat catalyze the...   \n",
       "484   l-serine dehydratase sdh, a member of the beta...   \n",
       "545   spermidine/spermine n1-acetyltransferase ssat ...   \n",
       "561   astrocytes may play a role in these manifestat...   \n",
       "607   reduced synthesis of nitric oxide no contribut...   \n",
       "661   estrogens are biosynthesised from androgens by...   \n",
       "682   background histamine synthesized by histidine ...   \n",
       "694   methionine synthase reductase mtrr is an enzym...   \n",
       "729   they all expressed ass, but not ornithine tran...   \n",
       "742   unlike mouse retsat mretsat, zretsat a had an ...   \n",
       "759   ldh is responsible for pyruvate conversion to ...   \n",
       "777   objective the aim was to test whether the comm...   \n",
       "789   in kidneys, stimulation of adenylyl cyclase ca...   \n",
       "881   adenylosuccinate synthetase adss catalyzes the...   \n",
       "905   risperidone is metabolized to its active metab...   \n",
       "995   aldo-keto reductases akrs metabolize a wide ra...   \n",
       "1086  aldose reductase ar catalyzes the reduction of...   \n",
       "1179  type 2 11beta-hydroxysteroid dehydrogenase enc...   \n",
       "1195  carboxylesterases hydrolyze esters, amides, an...   \n",
       "1367  pip5k1b encodes phosphatidylinositol 4-phospha...   \n",
       "1383  6-dhsg was metabolised by gsh to form a gsh co...   \n",
       "\n",
       "                                              chemicals  \n",
       "197                                   [glutamate, gaba]  \n",
       "252            [l-glutamate, gamma-glutamyl, l-proline]  \n",
       "282                                      [formate, ser]  \n",
       "318      [s-adenosylmethionine, s-adenosylhomocysteine]  \n",
       "323                         [glutamine, kynurenine, ka]  \n",
       "439     [capecitabine, thymidine, 5-fu, 5-fluorouracil]  \n",
       "465            [carnitine, acetyl-coa, acetylcarnitine]  \n",
       "484    [l-serine, 2-oxobutyrate, pyruvate, l-threonine]  \n",
       "545                   [spermidine, spermine, polyamine]  \n",
       "561                              [glutamate, glutamine]  \n",
       "607                      [l-arginine, no, nitric-oxide]  \n",
       "661                              [androgens, estrogens]  \n",
       "682                 [histidine, l-histidine, histamine]  \n",
       "694                                   [methionine, hcy]  \n",
       "729                   [citrulline, arginine, ornithine]  \n",
       "742   [all-trans-1314-dihydroretinol, all-trans-reti...  \n",
       "759                                 [pyruvate, lactate]  \n",
       "777                                         [ala, pufa]  \n",
       "789                              [amp, adenosine, camp]  \n",
       "881       [gtp, gdp, mg2+, adenylosuccinate, aspartate]  \n",
       "905                 [9-hydroxyrisperidone, risperidone]  \n",
       "995               [o-quinones, aldo-keto, oxygen, pahs]  \n",
       "1086                       [aldehydes, aldose, alcohol]  \n",
       "1179  [cortisone, cortisol, 11beta-hydroxysteroid, g...  \n",
       "1195  [amines, thiols, amides, alcohols, thioesters,...  \n",
       "1367             [phosphatidylinositol-45-bisphosphate]  \n",
       "1383                           [gs-6-dhsg, gsh, 6-dhsg]  "
      ]
     },
     "metadata": {},
     "execution_count": 329
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "### running test on the imported data"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "source": [
    "for index in range(len(sentence_chem_df[\"sentence\"])):\r\n",
    "    sentence = sentence_chem_df[\"sentence\"].iloc[index]\r\n",
    "    chems = sentence_chem_df[\"chemicals\"].iloc[index]\r\n",
    "    print(chems)\r\n",
    "    print(sentence)\r\n",
    "    print(\"general chemicals: \" + str(general_chemical(chems)))\r\n",
    "    print(\"adjacent mentions: \" + str(adjacent_mentions(chems, sentence)))\r\n",
    "    print(\"sep or: \" + str(sep_or(chems, sentence)))\r\n",
    "    print(\"sep and: \" + str(sep_and(chems, sentence)))\r\n",
    "    print(\"sep comma: \" + str(sep_comma(chems, sentence)))\r\n",
    "    print(\"sep via: \" + str(sep_via(chems, sentence)))\r\n",
    "    print(\"sep sym: \" + str(sep_sym(chems, sentence)))\r\n",
    "    print(\"followed ase: \" + str(followed_ase(chems, sentence)))\r\n",
    "    print(\"group: \" + str(group(sentence)))\r\n",
    "    print(\"followed by noun: \" + str(followed_by_noun(chems, sentence)))\r\n",
    "    print(\"sep verb: \" + str(sep_verb(chems, sentence)))\r\n",
    "    print(\"sep adverb: \" + str(sep_adverb(chems, sentence)))\r\n",
    "    print(\"includes reaction words: \" + str(includes_reaction_words(sentence)))\r\n",
    "    print()"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['glutamate', 'gaba']\n",
      "background and aims glutamic acid decarboxylase gad, ec 41115 catalyses the conversion of glutamate to gamma-aminobutyric acid gaba\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['l-glutamate', 'gamma-glutamyl', 'l-proline']\n",
      "interestingly, the allele of pro1 was shown to enhance the activities of gamma-glutamyl kinase and gamma-glutamyl phosphate reductase, both of which catalyze the first two steps of l-proline synthesis from l-glutamate and which together may form a complex in vivo\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "\n",
      "['formate', 'ser']\n",
      "we concluded that fdh has no direct role in the regulation of the above two pathways of ser synthesis the breakdown of formate to co2 by the fdh reaction is the primary and preferred fate of the organic acid in arabidopsis\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['s-adenosylmethionine', 's-adenosylhomocysteine']\n",
      "even though the activities of mat and gnmt were elevated, the concentration of liver s-adenosylmethionine was decreased 24%, p<0001 and s-adenosylhomocysteine increased 113%, p<0001 in the dwarf mice\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "\n",
      "['glutamine', 'kynurenine', 'ka']\n",
      "one of the enzymes responsible for the production of ka, kynurenine aminotransferase i kati, also catalyses the reversible transamination of glutamine to oxoglutaramic acid gtk, ec 26115\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: True\n",
      "\n",
      "['capecitabine', 'thymidine', '5-fu', '5-fluorouracil']\n",
      "purpose the fluoropyrimidine carbamate capecitabine is converted to 5-fluorouracil 5-fu by thymidine phosphorylase tp inside target tissues\n",
      "general chemicals: -1\n",
      "adjacent mentions: False\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: False\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['carnitine', 'acetyl-coa', 'acetylcarnitine']\n",
      "carnitine acetyltransferases crat catalyze the reversible conversion of acetyl-coa and carnitine to acetylcarnitine and free coa\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['l-serine', '2-oxobutyrate', 'pyruvate', 'l-threonine']\n",
      "l-serine dehydratase sdh, a member of the beta-family of pyridoxal phosphate-dependent plp enzymes, catalyzes the deamination of l-serine and l-threonine to yield pyruvate or 2-oxobutyrate\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: False\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "\n",
      "['spermidine', 'spermine', 'polyamine']\n",
      "spermidine/spermine n1-acetyltransferase ssat is a key enzyme in the control of polyamine levels in human cells, as acetylation of spermidine and spermine triggers export or degradation\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: False\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: False\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['glutamate', 'glutamine']\n",
      "astrocytes may play a role in these manifestations because astrocytes are essential in the regulation of released glutamate and its conversion to glutamine through the enzyme glutamine synthetase gs\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['l-arginine', 'no', 'nitric-oxide']\n",
      "reduced synthesis of nitric oxide no contributes to the endothelial dysfunction and may be related to limited availability of l-arginine, the common substrate of constitutive nitric-oxide synthase nos and cytosolic arginase i and mitochondrial arginase ii\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "\n",
      "['androgens', 'estrogens']\n",
      "estrogens are biosynthesised from androgens by the cyp450 enzyme complex called aromatase\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['histidine', 'l-histidine', 'histamine']\n",
      "background histamine synthesized by histidine decarboxylase hdc from l-histidine is a major chemical mediator in the development of nasal allergy which is characterized by nasal hypersensitivity\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['methionine', 'hcy']\n",
      "methionine synthase reductase mtrr is an enzyme involved in the conversion of hcy to methionine\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['citrulline', 'arginine', 'ornithine']\n",
      "they all expressed ass, but not ornithine transcarbamylase otc, the enzyme that converts ornithine, the product of degradation of arginine with rharg, to citrulline, which is converted back to arginine via ass\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: True\n",
      "\n",
      "['all-trans-1314-dihydroretinol', 'all-trans-retinol', 'all-trans-78-dihydroretinol']\n",
      "unlike mouse retsat mretsat, zretsat a had an altered bond specificity saturating either the 13-14 or 7-8 double bonds of all-trans-retinol to produce either all-trans-13,14-dihydroretinol or all-trans-7,8-dihydroretinol, respectively\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['pyruvate', 'lactate']\n",
      "ldh is responsible for pyruvate conversion to lactate through glycolysis\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['ala', 'pufa']\n",
      "objective the aim was to test whether the common deletion [t/-] in the promoter of fads2 affects the pufa biosynthetic pathway and consequently modifies the effect of alpha-linolenic acid ala on myocardial infarction mi\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: False\n",
      "includes reaction words: -1\n",
      "\n",
      "['amp', 'adenosine', 'camp']\n",
      "in kidneys, stimulation of adenylyl cyclase causes egress of camp, conversion of camp to amp by ecto-phosphodiesterase, and metabolism of amp to adenosine by ecto-5'-nucleotidase\n",
      "general chemicals: False\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['gtp', 'gdp', 'mg2+', 'adenylosuccinate', 'aspartate']\n",
      "adenylosuccinate synthetase adss catalyzes the mg2+ dependent condensation of a molecule of imp with aspartate to form adenylosuccinate, in a reaction driven by the hydrolysis of gtp to gdp\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['9-hydroxyrisperidone', 'risperidone']\n",
      "risperidone is metabolized to its active metabolite, 9-hydroxyrisperidone, mainly by the cytochrome p450 enzymes cyp2d6 and 3a4\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['o-quinones', 'aldo-keto', 'oxygen', 'pahs']\n",
      "aldo-keto reductases akrs metabolize a wide range of substrates, including polycyclic aromatic hydrocarbons pahs, generating metabolites o-quinones and reactive oxygen species ros, which are capable of initiating and promoting carcinogenesis\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['aldehydes', 'aldose', 'alcohol']\n",
      "aldose reductase ar catalyzes the reduction of toxic lipid aldehydes to their alcohol products and mediates inflammatory signals triggered by lipopolysaccharide lps\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: False\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['cortisone', 'cortisol', '11beta-hydroxysteroid', 'glucose']\n",
      "type 2 11beta-hydroxysteroid dehydrogenase encoded by the hsd11b2 gene converts cortisol to inactive cortisone, and alteration in this enzymatic activity might affect glucose homeostasis by affecting circulating levels or tissue availability of glucocorticoids\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: False\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['amines', 'thiols', 'amides', 'alcohols', 'thioesters', 'esters']\n",
      "carboxylesterases hydrolyze esters, amides, and thioesters to produce carboxylic acids and resulting alcohols, amines, and thiols, respectively\n",
      "general chemicals: False\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: False\n",
      "followed by noun: -1\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n",
      "['phosphatidylinositol-45-bisphosphate']\n",
      "pip5k1b encodes phosphatidylinositol 4-phosphate 5-kinase beta type i pip5k1beta, an enzyme functionally linked to actin cytoskeleton dynamics that phosphorylates phosphatidylinositol 4-phosphate [pi4p] to generate phosphatidylinositol-4,5-bisphosphate [pi4,5p2]\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: -1\n",
      "sep adverb: -1\n",
      "includes reaction words: -1\n",
      "\n",
      "['gs-6-dhsg', 'gsh', '6-dhsg']\n",
      "6-dhsg was metabolised by gsh to form a gsh conjugate gs-6-dhsg in raw 2647 cells, via a potential mechanism involving the catalytic activity of glutathione-s-transferase gst\n",
      "general chemicals: -1\n",
      "adjacent mentions: -1\n",
      "sep or: -1\n",
      "sep and: -1\n",
      "sep comma: -1\n",
      "sep via: -1\n",
      "sep sym: -1\n",
      "followed ase: -1\n",
      "group: -1\n",
      "followed by noun: False\n",
      "sep verb: True\n",
      "sep adverb: -1\n",
      "includes reaction words: True\n",
      "\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.9.7 64-bit (windows store)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "interpreter": {
   "hash": "00f65aa029e27b5bfb36b1f8c5b84bd40a443dbca505b0e94868b9845c12ad23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}